// allotment/dist/legacy.js
"use strict";

function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n4 = 0, F = function F() {}; return { s: F, n: function n() { return _n4 >= r.length ? { done: !0 } : { done: !1, value: r[_n4++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var e = require("react"),
  t = "undefined" != typeof globalThis ? globalThis : "undefined" != typeof window ? window : "undefined" != typeof global ? global : "undefined" != typeof self ? self : {};
function i(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e["default"] : e;
}
var s,
  n = {
    exports: {}
  };
/*!
	Copyright (c) 2018 Jed Watson.
	Licensed under the MIT License (MIT), see
	http://jedwatson.github.io/classnames
*/
var r,
  o,
  a,
  h = (s || (s = 1, r = n, function () {
    var e = {}.hasOwnProperty;
    function t() {
      for (var e = "", t = 0; t < arguments.length; t++) {
        var n = arguments[t];
        n && (e = s(e, i(n)));
      }
      return e;
    }
    function i(i) {
      if ("string" == typeof i || "number" == typeof i) return i;
      if ("object" != _typeof(i)) return "";
      if (Array.isArray(i)) return t.apply(null, i);
      if (i.toString !== Object.prototype.toString && !i.toString.toString().includes("[native code]")) return i.toString();
      var n = "";
      for (var r in i) e.call(i, r) && i[r] && (n = s(n, r));
      return n;
    }
    function s(e, t) {
      return t ? e ? e + " " + t : e + t : e;
    }
    r.exports ? (t["default"] = t, r.exports = t) : window.classNames = t;
  }()), n.exports),
  l = i(h);
var u,
  c,
  m = (a || (a = 1, o = function e(t, i) {
    if (t === i) return !0;
    if (t && i && "object" == _typeof(t) && "object" == _typeof(i)) {
      if (t.constructor !== i.constructor) return !1;
      var s, n, r;
      if (Array.isArray(t)) {
        if ((s = t.length) != i.length) return !1;
        for (n = s; 0 !== n--;) if (!e(t[n], i[n])) return !1;
        return !0;
      }
      if (t instanceof Map && i instanceof Map) {
        if (t.size !== i.size) return !1;
        var _iterator = _createForOfIteratorHelper(t.entries()),
          _step;
        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            n = _step.value;
            if (!i.has(n[0])) return !1;
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
        var _iterator2 = _createForOfIteratorHelper(t.entries()),
          _step2;
        try {
          for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
            n = _step2.value;
            if (!e(n[1], i.get(n[0]))) return !1;
          }
        } catch (err) {
          _iterator2.e(err);
        } finally {
          _iterator2.f();
        }
        return !0;
      }
      if (t instanceof Set && i instanceof Set) {
        if (t.size !== i.size) return !1;
        var _iterator3 = _createForOfIteratorHelper(t.entries()),
          _step3;
        try {
          for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
            n = _step3.value;
            if (!i.has(n[0])) return !1;
          }
        } catch (err) {
          _iterator3.e(err);
        } finally {
          _iterator3.f();
        }
        return !0;
      }
      if (ArrayBuffer.isView(t) && ArrayBuffer.isView(i)) {
        if ((s = t.length) != i.length) return !1;
        for (n = s; 0 !== n--;) if (t[n] !== i[n]) return !1;
        return !0;
      }
      if (t.constructor === RegExp) return t.source === i.source && t.flags === i.flags;
      if (t.valueOf !== Object.prototype.valueOf) return t.valueOf() === i.valueOf();
      if (t.toString !== Object.prototype.toString) return t.toString() === i.toString();
      if ((s = (r = Object.keys(t)).length) !== Object.keys(i).length) return !1;
      for (n = s; 0 !== n--;) if (!Object.prototype.hasOwnProperty.call(i, r[n])) return !1;
      for (n = s; 0 !== n--;) {
        var o = r[n];
        if (!e(t[o], i[o])) return !1;
      }
      return !0;
    }
    return t != t && i != i;
  }), o),
  f = i(m);
var d,
  p,
  v = i(function () {
    if (c) return u;
    c = 1;
    var e = /^\s+|\s+$/g,
      t = /^[-+]0x[0-9a-f]+$/i,
      i = /^0b[01]+$/i,
      s = /^0o[0-7]+$/i,
      n = parseInt,
      r = Object.prototype.toString;
    function o(e) {
      var t = _typeof(e);
      return !!e && ("object" == t || "function" == t);
    }
    function a(a) {
      if ("number" == typeof a) return a;
      if (function (e) {
        return "symbol" == _typeof(e) || function (e) {
          return !!e && "object" == _typeof(e);
        }(e) && "[object Symbol]" == r.call(e);
      }(a)) return NaN;
      if (o(a)) {
        var h = "function" == typeof a.valueOf ? a.valueOf() : a;
        a = o(h) ? h + "" : h;
      }
      if ("string" != typeof a) return 0 === a ? a : +a;
      a = a.replace(e, "");
      var l = i.test(a);
      return l || s.test(a) ? n(a.slice(2), l ? 2 : 8) : t.test(a) ? NaN : +a;
    }
    return u = function u(e, t, i) {
      return void 0 === i && (i = t, t = void 0), void 0 !== i && (i = (i = a(i)) == i ? i : 0), void 0 !== t && (t = (t = a(t)) == t ? t : 0), function (e, t, i) {
        return e == e && (void 0 !== i && (e = e <= i ? e : i), void 0 !== t && (e = e >= t ? e : t)), e;
      }(a(e), t, i);
    };
  }());
var S = i(function () {
  if (p) return d;
  p = 1;
  var e = /^\s+|\s+$/g,
    i = /^[-+]0x[0-9a-f]+$/i,
    s = /^0b[01]+$/i,
    n = /^0o[0-7]+$/i,
    r = parseInt,
    o = "object" == _typeof(t) && t && t.Object === Object && t,
    a = "object" == (typeof self === "undefined" ? "undefined" : _typeof(self)) && self && self.Object === Object && self,
    h = o || a || Function("return this")(),
    l = Object.prototype.toString,
    u = Math.max,
    c = Math.min,
    m = function m() {
      return h.Date.now();
    };
  function f(e) {
    var t = _typeof(e);
    return !!e && ("object" == t || "function" == t);
  }
  function v(t) {
    if ("number" == typeof t) return t;
    if (function (e) {
      return "symbol" == _typeof(e) || function (e) {
        return !!e && "object" == _typeof(e);
      }(e) && "[object Symbol]" == l.call(e);
    }(t)) return NaN;
    if (f(t)) {
      var o = "function" == typeof t.valueOf ? t.valueOf() : t;
      t = f(o) ? o + "" : o;
    }
    if ("string" != typeof t) return 0 === t ? t : +t;
    t = t.replace(e, "");
    var a = s.test(t);
    return a || n.test(t) ? r(t.slice(2), a ? 2 : 8) : i.test(t) ? NaN : +t;
  }
  return d = function d(e, t, i) {
    var s,
      n,
      r,
      o,
      a,
      h,
      l = 0,
      d = !1,
      p = !1,
      S = !0;
    if ("function" != typeof e) throw new TypeError("Expected a function");
    function z(t) {
      var i = s,
        r = n;
      return s = n = void 0, l = t, o = e.apply(r, i);
    }
    function w(e) {
      var i = e - h;
      return void 0 === h || i >= t || i < 0 || p && e - l >= r;
    }
    function g() {
      var e = m();
      if (w(e)) return y(e);
      a = setTimeout(g, function (e) {
        var i = t - (e - h);
        return p ? c(i, r - (e - l)) : i;
      }(e));
    }
    function y(e) {
      return a = void 0, S && s ? z(e) : (s = n = void 0, o);
    }
    function b() {
      var e = m(),
        i = w(e);
      if (s = arguments, n = this, h = e, i) {
        if (void 0 === a) return function (e) {
          return l = e, a = setTimeout(g, t), d ? z(e) : o;
        }(h);
        if (p) return a = setTimeout(g, t), z(h);
      }
      return void 0 === a && (a = setTimeout(g, t)), o;
    }
    return t = v(t) || 0, f(i) && (d = !!i.leading, r = (p = "maxWait" in i) ? u(v(i.maxWait) || 0, t) : r, S = "trailing" in i ? !!i.trailing : S), b.cancel = function () {
      void 0 !== a && clearTimeout(a), l = 0, s = h = n = a = void 0;
    }, b.flush = function () {
      return void 0 === a ? o : y(m());
    }, b;
  };
}());
var z = {
  width: void 0,
  height: void 0
};
function w(t) {
  var i = t.ref,
    _t$box = t.box,
    s = _t$box === void 0 ? "content-box" : _t$box,
    _e$useState = e.useState(z),
    _e$useState2 = _slicedToArray(_e$useState, 2),
    _e$useState2$ = _e$useState2[0],
    n = _e$useState2$.width,
    r = _e$useState2$.height,
    o = _e$useState2[1],
    a = function () {
      var t = e.useRef(!1);
      return e.useEffect(function () {
        return t.current = !0, function () {
          t.current = !1;
        };
      }, []), e.useCallback(function () {
        return t.current;
      }, []);
    }(),
    h = e.useRef(_objectSpread({}, z)),
    l = e.useRef(void 0);
  return l.current = t.onResize, e.useEffect(function () {
    if (!i.current) return;
    if ("undefined" == typeof window || !("ResizeObserver" in window)) return;
    var e = new ResizeObserver(function (_ref) {
      var _ref2 = _slicedToArray(_ref, 1),
        e = _ref2[0];
      var t = "border-box" === s ? "borderBoxSize" : "device-pixel-content-box" === s ? "devicePixelContentBoxSize" : "contentBoxSize",
        i = g(e, t, "inlineSize"),
        n = g(e, t, "blockSize");
      if (h.current.width !== i || h.current.height !== n) {
        var _e = {
          width: i,
          height: n
        };
        h.current.width = i, h.current.height = n, l.current ? l.current(_e) : a() && o(_e);
      }
    });
    return e.observe(i.current, {
      box: s
    }), function () {
      e.disconnect();
    };
  }, [s, i, a]), {
    width: n,
    height: r
  };
}
function g(e, t, i) {
  return e[t] ? Array.isArray(e[t]) ? e[t][0][i] : e[t][i] : "contentBoxSize" === t ? e.contentRect["inlineSize" === i ? "width" : "height"] : void 0;
}
var y = "allotment-module_splitView__L-yRc",
  b = "allotment-module_sashContainer__fzwJF",
  I = "allotment-module_splitViewContainer__rQnVa",
  x = "allotment-module_splitViewView__MGZ6O",
  _ = "allotment-module_vertical__WSwwa",
  E = "allotment-module_horizontal__7doS8",
  V = "allotment-module_separatorBorder__x-rDS";
var N,
  L = !1,
  D = !1;
"object" == (typeof navigator === "undefined" ? "undefined" : _typeof(navigator)) && (N = navigator.userAgent, D = N.indexOf("Macintosh") >= 0, L = (N.indexOf("Macintosh") >= 0 || N.indexOf("iPad") >= 0 || N.indexOf("iPhone") >= 0) && !!navigator.maxTouchPoints && navigator.maxTouchPoints > 0);
var M = L,
  O = D,
  P = "undefined" != typeof window && void 0 !== window.document && void 0 !== window.document.createElement ? e.useLayoutEffect : e.useEffect;
var C = /*#__PURE__*/function () {
  function C() {
    _classCallCheck(this, C);
    this._size = void 0;
  }
  return _createClass(C, [{
    key: "getSize",
    value: function getSize() {
      return this._size;
    }
  }, {
    key: "setSize",
    value: function setSize(e) {
      this._size = e;
    }
  }]);
}();
function T(e, t) {
  var i = e.length,
    s = i - t.length;
  return s >= 0 && e.slice(s, i) === t;
}
var A,
  j = {
    exports: {}
  };
var F = (A || (A = 1, function (e) {
    var t = Object.prototype.hasOwnProperty,
      i = "~";
    function s() {}
    function n(e, t, i) {
      this.fn = e, this.context = t, this.once = i || !1;
    }
    function r(e, t, s, r, o) {
      if ("function" != typeof s) throw new TypeError("The listener must be a function");
      var a = new n(s, r || e, o),
        h = i ? i + t : t;
      return e._events[h] ? e._events[h].fn ? e._events[h] = [e._events[h], a] : e._events[h].push(a) : (e._events[h] = a, e._eventsCount++), e;
    }
    function o(e, t) {
      0 === --e._eventsCount ? e._events = new s() : delete e._events[t];
    }
    function a() {
      this._events = new s(), this._eventsCount = 0;
    }
    Object.create && (s.prototype = Object.create(null), new s().__proto__ || (i = !1)), a.prototype.eventNames = function () {
      var e,
        s,
        n = [];
      if (0 === this._eventsCount) return n;
      for (s in e = this._events) t.call(e, s) && n.push(i ? s.slice(1) : s);
      return Object.getOwnPropertySymbols ? n.concat(Object.getOwnPropertySymbols(e)) : n;
    }, a.prototype.listeners = function (e) {
      var t = i ? i + e : e,
        s = this._events[t];
      if (!s) return [];
      if (s.fn) return [s.fn];
      for (var n = 0, r = s.length, o = new Array(r); n < r; n++) o[n] = s[n].fn;
      return o;
    }, a.prototype.listenerCount = function (e) {
      var t = i ? i + e : e,
        s = this._events[t];
      return s ? s.fn ? 1 : s.length : 0;
    }, a.prototype.emit = function (e, t, s, n, r, o) {
      var a = i ? i + e : e;
      if (!this._events[a]) return !1;
      var h,
        l,
        u = this._events[a],
        c = arguments.length;
      if (u.fn) {
        switch (u.once && this.removeListener(e, u.fn, void 0, !0), c) {
          case 1:
            return u.fn.call(u.context), !0;
          case 2:
            return u.fn.call(u.context, t), !0;
          case 3:
            return u.fn.call(u.context, t, s), !0;
          case 4:
            return u.fn.call(u.context, t, s, n), !0;
          case 5:
            return u.fn.call(u.context, t, s, n, r), !0;
          case 6:
            return u.fn.call(u.context, t, s, n, r, o), !0;
        }
        for (l = 1, h = new Array(c - 1); l < c; l++) h[l - 1] = arguments[l];
        u.fn.apply(u.context, h);
      } else {
        var m,
          f = u.length;
        for (l = 0; l < f; l++) switch (u[l].once && this.removeListener(e, u[l].fn, void 0, !0), c) {
          case 1:
            u[l].fn.call(u[l].context);
            break;
          case 2:
            u[l].fn.call(u[l].context, t);
            break;
          case 3:
            u[l].fn.call(u[l].context, t, s);
            break;
          case 4:
            u[l].fn.call(u[l].context, t, s, n);
            break;
          default:
            if (!h) for (m = 1, h = new Array(c - 1); m < c; m++) h[m - 1] = arguments[m];
            u[l].fn.apply(u[l].context, h);
        }
      }
      return !0;
    }, a.prototype.on = function (e, t, i) {
      return r(this, e, t, i, !1);
    }, a.prototype.once = function (e, t, i) {
      return r(this, e, t, i, !0);
    }, a.prototype.removeListener = function (e, t, s, n) {
      var r = i ? i + e : e;
      if (!this._events[r]) return this;
      if (!t) return o(this, r), this;
      var a = this._events[r];
      if (a.fn) a.fn !== t || n && !a.once || s && a.context !== s || o(this, r);else {
        for (var h = 0, l = [], u = a.length; h < u; h++) (a[h].fn !== t || n && !a[h].once || s && a[h].context !== s) && l.push(a[h]);
        l.length ? this._events[r] = 1 === l.length ? l[0] : l : o(this, r);
      }
      return this;
    }, a.prototype.removeAllListeners = function (e) {
      var t;
      return e ? (t = i ? i + e : e, this._events[t] && o(this, t)) : (this._events = new s(), this._eventsCount = 0), this;
    }, a.prototype.off = a.prototype.removeListener, a.prototype.addListener = a.prototype.on, a.prefixed = i, a.EventEmitter = a, e.exports = a;
  }(j)), j.exports),
  H = i(F);
function R(e, t) {
  var i = e.indexOf(t);
  i > -1 && (e.splice(i, 1), e.unshift(t));
}
function k(e, t) {
  var i = e.indexOf(t);
  i > -1 && (e.splice(i, 1), e.push(t));
}
function Y(e, t) {
  var i = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 1;
  var s = Math.max(0, Math.ceil((t - e) / i)),
    n = new Array(s);
  var r = -1;
  for (; ++r < s;) n[r] = e + r * i;
  return n;
}
var $ = "sash-module_sash__K-9lB",
  B = "sash-module_disabled__Hm-wx",
  W = "sash-module_mac__Jf6OJ",
  X = "sash-module_vertical__pB-rs",
  G = "sash-module_minimum__-UKxp",
  J = "sash-module_maximum__TCWxD",
  U = "sash-module_horizontal__kFbiw",
  K = "sash-module_hover__80W6I",
  Z = "sash-module_active__bJspD";
var q = function (e) {
    return e.Vertical = "VERTICAL", e.Horizontal = "HORIZONTAL", e;
  }({}),
  Q = function (e) {
    return e.Disabled = "DISABLED", e.Minimum = "MINIMUM", e.Maximum = "MAXIMUM", e.Enabled = "ENABLED", e;
  }({}),
  ee = M ? 20 : 8;
var te = new H();
var ie = /*#__PURE__*/function (_H) {
  function ie(e, t, i) {
    var _i$orientation;
    var _this;
    _classCallCheck(this, ie);
    _this = _callSuper(this, ie), _this.el = void 0, _this.layoutProvider = void 0, _this.orientation = void 0, _this.size = void 0, _this.hoverDelay = 300, _this.hoverDelayer = S(function (e) {
      return e.classList.add("sash-hover", K);
    }, _this.hoverDelay), _this._state = Q.Enabled, _this.onPointerStart = function (e) {
      var t = e.pageX,
        i = e.pageY,
        s = {
          startX: t,
          currentX: t,
          startY: i,
          currentY: i
        };
      _this.el.classList.add("sash-active", Z), _this.emit("start", s), _this.el.setPointerCapture(e.pointerId);
      var n = function n(e) {
          e.preventDefault();
          var s = {
            startX: t,
            currentX: e.pageX,
            startY: i,
            currentY: e.pageY
          };
          _this.emit("change", s);
        },
        _r = function r(e) {
          e.preventDefault(), _this.el.classList.remove("sash-active", Z), _this.hoverDelayer.cancel(), _this.emit("end"), _this.el.releasePointerCapture(e.pointerId), window.removeEventListener("pointermove", n), window.removeEventListener("pointerup", _r);
        };
      window.addEventListener("pointermove", n), window.addEventListener("pointerup", _r);
    }, _this.onPointerDoublePress = function () {
      _this.emit("reset");
    }, _this.onMouseEnter = function () {
      _this.el.classList.contains(Z) ? (_this.hoverDelayer.cancel(), _this.el.classList.add("sash-hover", K)) : _this.hoverDelayer(_this.el);
    }, _this.onMouseLeave = function () {
      _this.hoverDelayer.cancel(), _this.el.classList.remove("sash-hover", K);
    }, _this.el = document.createElement("div"), _this.el.classList.add("sash", $), _this.el.dataset.testid = "sash", e.append(_this.el), O && _this.el.classList.add("sash-mac", W), _this.el.addEventListener("pointerdown", _this.onPointerStart), _this.el.addEventListener("dblclick", _this.onPointerDoublePress), _this.el.addEventListener("mouseenter", _this.onMouseEnter), _this.el.addEventListener("mouseleave", _this.onMouseLeave), "number" == typeof i.size ? (_this.size = i.size, i.orientation === q.Vertical ? _this.el.style.width = "".concat(_this.size, "px") : _this.el.style.height = "".concat(_this.size, "px")) : (_this.size = ee, te.on("onDidChangeGlobalSize", function (e) {
      _this.size = e, _this.layout();
    })), _this.layoutProvider = t, _this.orientation = (_i$orientation = i.orientation) !== null && _i$orientation !== void 0 ? _i$orientation : q.Vertical, _this.orientation === q.Horizontal ? (_this.el.classList.add("sash-horizontal", U), _this.el.classList.remove("sash-vertical", X)) : (_this.el.classList.remove("sash-horizontal", U), _this.el.classList.add("sash-vertical", X)), _this.layout();
    return _this;
  }
  _inherits(ie, _H);
  return _createClass(ie, [{
    key: "state",
    get: function get() {
      return this._state;
    },
    set: function set(e) {
      this._state !== e && (this.el.classList.toggle(B, e === Q.Disabled), this.el.classList.toggle("sash-disabled", e === Q.Disabled), this.el.classList.toggle(G, e === Q.Minimum), this.el.classList.toggle("sash-minimum", e === Q.Minimum), this.el.classList.toggle(J, e === Q.Maximum), this.el.classList.toggle("sash-maximum", e === Q.Maximum), this._state = e, this.emit("enablementChange", e));
    }
  }, {
    key: "layout",
    value: function layout() {
      if (this.orientation === q.Vertical) {
        var _e2 = this.layoutProvider;
        this.el.style.left = _e2.getVerticalSashLeft(this) - this.size / 2 + "px", _e2.getVerticalSashTop && (this.el.style.top = _e2.getVerticalSashTop(this) + "px"), _e2.getVerticalSashHeight && (this.el.style.height = _e2.getVerticalSashHeight(this) + "px");
      } else {
        var _e3 = this.layoutProvider;
        this.el.style.top = _e3.getHorizontalSashTop(this) - this.size / 2 + "px", _e3.getHorizontalSashLeft && (this.el.style.left = _e3.getHorizontalSashLeft(this) + "px"), _e3.getHorizontalSashWidth && (this.el.style.width = _e3.getHorizontalSashWidth(this) + "px");
      }
    }
  }, {
    key: "dispose",
    value: function dispose() {
      var _this2 = this;
      this.el.removeEventListener("pointerdown", this.onPointerStart), this.el.removeEventListener("dblclick", this.onPointerDoublePress), this.el.removeEventListener("mouseenter", this.onMouseEnter), this.el.removeEventListener("mouseleave", function () {
        return _this2.onMouseLeave;
      }), this.el.remove();
    }
  }]);
}(H);
var se;
var ne;
(ne = se || (se = {})).Distribute = {
  type: "distribute"
}, ne.Split = function (e) {
  return {
    type: "split",
    index: e
  };
}, ne.Invisible = function (e) {
  return {
    type: "invisible",
    cachedVisibleSize: e
  };
};
var re = function (e) {
  return e.Normal = "NORMAL", e.Low = "LOW", e.High = "HIGH", e;
}({});
var oe = /*#__PURE__*/function () {
  function oe(e, t, i) {
    _classCallCheck(this, oe);
    this.container = void 0, this.view = void 0, this._size = void 0, this._cachedVisibleSize = void 0, this.container = e, this.view = t, this.container.classList.add("split-view-view", x), this.container.dataset.testid = "split-view-view", "number" == typeof i ? (this._size = i, this._cachedVisibleSize = void 0, e.classList.add("split-view-view-visible")) : (this._size = 0, this._cachedVisibleSize = i.cachedVisibleSize);
  }
  return _createClass(oe, [{
    key: "size",
    get: function get() {
      return this._size;
    },
    set: function set(e) {
      this._size = e;
    }
  }, {
    key: "priority",
    get: function get() {
      return this.view.priority;
    }
  }, {
    key: "snap",
    get: function get() {
      return !!this.view.snap;
    }
  }, {
    key: "cachedVisibleSize",
    get: function get() {
      return this._cachedVisibleSize;
    }
  }, {
    key: "visible",
    get: function get() {
      return void 0 === this._cachedVisibleSize;
    }
  }, {
    key: "setVisible",
    value: function setVisible(e, t) {
      e !== this.visible && (e ? (this.size = v(this._cachedVisibleSize, this.viewMinimumSize, this.viewMaximumSize), this._cachedVisibleSize = void 0) : (this._cachedVisibleSize = "number" == typeof t ? t : this.size, this.size = 0), this.container.classList.toggle("split-view-view-visible", e), this.view.setVisible && this.view.setVisible(e));
    }
  }, {
    key: "minimumSize",
    get: function get() {
      return this.visible ? this.view.minimumSize : 0;
    }
  }, {
    key: "viewMinimumSize",
    get: function get() {
      return this.view.minimumSize;
    }
  }, {
    key: "maximumSize",
    get: function get() {
      return this.visible ? this.view.maximumSize : 0;
    }
  }, {
    key: "viewMaximumSize",
    get: function get() {
      return this.view.maximumSize;
    }
  }, {
    key: "enabled",
    set: function set(e) {
      this.container.style.pointerEvents = e ? "" : "none";
    }
  }, {
    key: "layout",
    value: function layout(e) {
      this.layoutContainer(e), this.view.layout(this.size, e);
    }
  }]);
}();
var ae = /*#__PURE__*/function (_oe) {
  function ae() {
    _classCallCheck(this, ae);
    return _callSuper(this, ae, arguments);
  }
  _inherits(ae, _oe);
  return _createClass(ae, [{
    key: "layoutContainer",
    value: function layoutContainer(e) {
      this.container.style.left = "".concat(e, "px"), this.container.style.width = "".concat(this.size, "px");
    }
  }]);
}(oe);
var he = /*#__PURE__*/function (_oe2) {
  function he() {
    _classCallCheck(this, he);
    return _callSuper(this, he, arguments);
  }
  _inherits(he, _oe2);
  return _createClass(he, [{
    key: "layoutContainer",
    value: function layoutContainer(e) {
      this.container.style.top = "".concat(e, "px"), this.container.style.height = "".concat(this.size, "px");
    }
  }]);
}(oe);
var le = /*#__PURE__*/function (_H2) {
  function le(e) {
    var _t$orientation, _t$proportionalLayout;
    var _this3;
    var t = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var i = arguments.length > 2 ? arguments[2] : undefined;
    var s = arguments.length > 3 ? arguments[3] : undefined;
    var n = arguments.length > 4 ? arguments[4] : undefined;
    _classCallCheck(this, le);
    if (_this3 = _callSuper(this, le), _this3.onDidChange = void 0, _this3.onDidDragStart = void 0, _this3.onDidDragEnd = void 0, _this3.orientation = void 0, _this3.sashContainer = void 0, _this3.size = 0, _this3.contentSize = 0, _this3.proportions = void 0, _this3.viewItems = [], _this3.sashItems = [], _this3.sashDragState = void 0, _this3.proportionalLayout = void 0, _this3.getSashOrthogonalSize = void 0, _this3._startSnappingEnabled = !0, _this3._endSnappingEnabled = !0, _this3.onSashEnd = function (e) {
      _this3.emit("sashchange", e), _this3.saveProportions();
      var _iterator4 = _createForOfIteratorHelper(_this3.viewItems),
        _step4;
      try {
        for (_iterator4.s(); !(_step4 = _iterator4.n()).done;) {
          var _e4 = _step4.value;
          _e4.enabled = !0;
        }
      } catch (err) {
        _iterator4.e(err);
      } finally {
        _iterator4.f();
      }
    }, _this3.orientation = (_t$orientation = t.orientation) !== null && _t$orientation !== void 0 ? _t$orientation : q.Vertical, _this3.proportionalLayout = (_t$proportionalLayout = t.proportionalLayout) !== null && _t$proportionalLayout !== void 0 ? _t$proportionalLayout : !0, _this3.getSashOrthogonalSize = t.getSashOrthogonalSize, i && (_this3.onDidChange = i), s && (_this3.onDidDragStart = s), n && (_this3.onDidDragEnd = n), _this3.sashContainer = document.createElement("div"), _this3.sashContainer.classList.add("sash-container", b), e.prepend(_this3.sashContainer), t.descriptor) {
      _this3.size = t.descriptor.size;
      var _iterator5 = _createForOfIteratorHelper(t.descriptor.views.entries()),
        _step5;
      try {
        for (_iterator5.s(); !(_step5 = _iterator5.n()).done;) {
          var _step5$value = _slicedToArray(_step5.value, 2),
            _e5 = _step5$value[0],
            _i = _step5$value[1];
          var _t = _i.size,
            _s = _i.container,
            _n = _i.view;
          _this3.addView(_s, _n, _t, _e5, !0);
        }
      } catch (err) {
        _iterator5.e(err);
      } finally {
        _iterator5.f();
      }
      _this3.contentSize = _this3.viewItems.reduce(function (e, t) {
        return e + t.size;
      }, 0), _this3.saveProportions();
    }
    return _this3;
  }
  _inherits(le, _H2);
  return _createClass(le, [{
    key: "startSnappingEnabled",
    get: function get() {
      return this._startSnappingEnabled;
    },
    set: function set(e) {
      this._startSnappingEnabled !== e && (this._startSnappingEnabled = e, this.updateSashEnablement());
    }
  }, {
    key: "endSnappingEnabled",
    get: function get() {
      return this._endSnappingEnabled;
    },
    set: function set(e) {
      this._endSnappingEnabled !== e && (this._endSnappingEnabled = e, this.updateSashEnablement());
    }
  }, {
    key: "addView",
    value: function addView(e, t, i) {
      var _this4 = this;
      var s = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : this.viewItems.length;
      var n = arguments.length > 4 ? arguments[4] : undefined;
      var r;
      r = "number" == typeof i ? i : "split" === i.type ? this.getViewSize(i.index) / 2 : "invisible" === i.type ? {
        cachedVisibleSize: i.cachedVisibleSize
      } : t.minimumSize;
      var o = this.orientation === q.Vertical ? new he(e, t, r) : new ae(e, t, r);
      if (this.viewItems.splice(s, 0, o), this.viewItems.length > 1) {
        var _e6 = this.orientation === q.Vertical ? new ie(this.sashContainer, {
            getHorizontalSashTop: function getHorizontalSashTop(e) {
              return _this4.getSashPosition(e);
            },
            getHorizontalSashWidth: this.getSashOrthogonalSize
          }, {
            orientation: q.Horizontal
          }) : new ie(this.sashContainer, {
            getVerticalSashLeft: function getVerticalSashLeft(e) {
              return _this4.getSashPosition(e);
            },
            getVerticalSashHeight: this.getSashOrthogonalSize
          }, {
            orientation: q.Vertical
          }),
          _t2 = this.orientation === q.Vertical ? function (t) {
            return {
              sash: _e6,
              start: t.startY,
              current: t.currentY
            };
          } : function (t) {
            return {
              sash: _e6,
              start: t.startX,
              current: t.currentX
            };
          };
        _e6.on("start", function (e) {
          var _this4$onDidDragStart;
          _this4.emit("sashDragStart"), _this4.onSashStart(_t2(e));
          var i = _this4.viewItems.map(function (e) {
            return e.size;
          });
          (_this4$onDidDragStart = _this4.onDidDragStart) === null || _this4$onDidDragStart === void 0 || _this4$onDidDragStart.call(_this4, i);
        }), _e6.on("change", function (e) {
          return _this4.onSashChange(_t2(e));
        }), _e6.on("end", function () {
          var _this4$onDidDragEnd;
          _this4.emit("sashDragEnd"), _this4.onSashEnd(_this4.sashItems.findIndex(function (t) {
            return t.sash === _e6;
          }));
          var t = _this4.viewItems.map(function (e) {
            return e.size;
          });
          (_this4$onDidDragEnd = _this4.onDidDragEnd) === null || _this4$onDidDragEnd === void 0 || _this4$onDidDragEnd.call(_this4, t);
        }), _e6.on("reset", function () {
          var t = _this4.sashItems.findIndex(function (t) {
              return t.sash === _e6;
            }),
            i = Y(t, -1, -1),
            s = Y(t + 1, _this4.viewItems.length),
            n = _this4.findFirstSnapIndex(i),
            r = _this4.findFirstSnapIndex(s);
          ("number" != typeof n || _this4.viewItems[n].visible) && ("number" != typeof r || _this4.viewItems[r].visible) && _this4.emit("sashreset", t);
        });
        var _i2 = {
          sash: _e6
        };
        this.sashItems.splice(s - 1, 0, _i2);
      }
      n || this.relayout(), n || "number" == typeof i || "distribute" !== i.type || this.distributeViewSizes();
    }
  }, {
    key: "removeView",
    value: function removeView(e, t) {
      if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
      var i = this.viewItems.splice(e, 1)[0].view;
      if (this.viewItems.length >= 1) {
        var _t3 = Math.max(e - 1, 0);
        this.sashItems.splice(_t3, 1)[0].sash.dispose();
      }
      return this.relayout(), t && "distribute" === t.type && this.distributeViewSizes(), i;
    }
  }, {
    key: "moveView",
    value: function moveView(e, t, i) {
      var s = this.getViewCachedVisibleSize(t),
        n = void 0 === s ? this.getViewSize(t) : se.Invisible(s),
        r = this.removeView(t);
      this.addView(e, r, n, i);
    }
  }, {
    key: "getViewCachedVisibleSize",
    value: function getViewCachedVisibleSize(e) {
      if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
      return this.viewItems[e].cachedVisibleSize;
    }
  }, {
    key: "layout",
    value: function layout() {
      var _this5 = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.size;
      var t = Math.max(this.size, this.contentSize);
      if (this.size = e, this.proportions) for (var _t4 = 0; _t4 < this.viewItems.length; _t4++) {
        var _i3 = this.viewItems[_t4];
        _i3.size = v(Math.round(this.proportions[_t4] * e), _i3.minimumSize, _i3.maximumSize);
      } else {
        var _i4 = Y(0, this.viewItems.length),
          _s2 = _i4.filter(function (e) {
            return _this5.viewItems[e].priority === re.Low;
          }),
          _n2 = _i4.filter(function (e) {
            return _this5.viewItems[e].priority === re.High;
          });
        this.resize(this.viewItems.length - 1, e - t, void 0, _s2, _n2);
      }
      this.distributeEmptySpace(), this.layoutViews();
    }
  }, {
    key: "resizeView",
    value: function resizeView(e, t) {
      var _this6 = this;
      if (e < 0 || e >= this.viewItems.length) return;
      var i = Y(0, this.viewItems.length).filter(function (t) {
          return t !== e;
        }),
        s = [].concat(_toConsumableArray(i.filter(function (e) {
          return _this6.viewItems[e].priority === re.Low;
        })), [e]),
        n = i.filter(function (e) {
          return _this6.viewItems[e].priority === re.High;
        }),
        r = this.viewItems[e];
      t = Math.round(t), t = v(t, r.minimumSize, Math.min(r.maximumSize, this.size)), r.size = t, this.relayout(s, n);
    }
  }, {
    key: "resizeViews",
    value: function resizeViews(e) {
      for (var _t5 = 0; _t5 < e.length; _t5++) {
        var _i5 = this.viewItems[_t5];
        var _s3 = e[_t5];
        _s3 = Math.round(_s3), _s3 = v(_s3, _i5.minimumSize, Math.min(_i5.maximumSize, this.size)), _i5.size = _s3;
      }
      this.contentSize = this.viewItems.reduce(function (e, t) {
        return e + t.size;
      }, 0), this.saveProportions(), this.layout(this.size);
    }
  }, {
    key: "getViewSize",
    value: function getViewSize(e) {
      return e < 0 || e >= this.viewItems.length ? -1 : this.viewItems[e].size;
    }
  }, {
    key: "isViewVisible",
    value: function isViewVisible(e) {
      if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
      return this.viewItems[e].visible;
    }
  }, {
    key: "setViewVisible",
    value: function setViewVisible(e, t) {
      if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
      this.viewItems[e].setVisible(t), this.distributeEmptySpace(e), this.layoutViews(), this.saveProportions();
    }
  }, {
    key: "distributeViewSizes",
    value: function distributeViewSizes() {
      var _this7 = this;
      var e = [];
      var t = 0;
      var _iterator6 = _createForOfIteratorHelper(this.viewItems),
        _step6;
      try {
        for (_iterator6.s(); !(_step6 = _iterator6.n()).done;) {
          var _i7 = _step6.value;
          _i7.maximumSize - _i7.minimumSize > 0 && (e.push(_i7), t += _i7.size);
        }
      } catch (err) {
        _iterator6.e(err);
      } finally {
        _iterator6.f();
      }
      var i = Math.floor(t / e.length);
      for (var _i6 = 0, _e7 = e; _i6 < _e7.length; _i6++) {
        var _t6 = _e7[_i6];
        _t6.size = v(i, _t6.minimumSize, _t6.maximumSize);
      }
      var s = Y(0, this.viewItems.length),
        n = s.filter(function (e) {
          return _this7.viewItems[e].priority === re.Low;
        }),
        r = s.filter(function (e) {
          return _this7.viewItems[e].priority === re.High;
        });
      this.relayout(n, r);
    }
  }, {
    key: "dispose",
    value: function dispose() {
      this.sashItems.forEach(function (e) {
        return e.sash.dispose();
      }), this.sashItems = [], this.sashContainer.remove();
    }
  }, {
    key: "relayout",
    value: function relayout(e, t) {
      var i = this.viewItems.reduce(function (e, t) {
        return e + t.size;
      }, 0);
      this.resize(this.viewItems.length - 1, this.size - i, void 0, e, t), this.distributeEmptySpace(), this.layoutViews(), this.saveProportions();
    }
  }, {
    key: "onSashStart",
    value: function onSashStart(_ref3) {
      var _this8 = this;
      var e = _ref3.sash,
        t = _ref3.start;
      var i = this.sashItems.findIndex(function (t) {
        return t.sash === e;
      });
      (function (e) {
        var t = _this8.viewItems.map(function (e) {
          return e.size;
        });
        var s,
          n,
          r = Number.NEGATIVE_INFINITY,
          o = Number.POSITIVE_INFINITY;
        var a = Y(i, -1, -1),
          h = Y(i + 1, _this8.viewItems.length),
          l = a.reduce(function (e, i) {
            return e + (_this8.viewItems[i].minimumSize - t[i]);
          }, 0),
          u = a.reduce(function (e, i) {
            return e + (_this8.viewItems[i].viewMaximumSize - t[i]);
          }, 0),
          c = 0 === h.length ? Number.POSITIVE_INFINITY : h.reduce(function (e, i) {
            return e + (t[i] - _this8.viewItems[i].minimumSize);
          }, 0),
          m = 0 === h.length ? Number.NEGATIVE_INFINITY : h.reduce(function (e, i) {
            return e + (t[i] - _this8.viewItems[i].viewMaximumSize);
          }, 0);
        r = Math.max(l, m), o = Math.min(c, u);
        var f = _this8.findFirstSnapIndex(a),
          d = _this8.findFirstSnapIndex(h);
        if ("number" == typeof f) {
          var _e8 = _this8.viewItems[f],
            _t7 = Math.floor(_e8.viewMinimumSize / 2);
          s = {
            index: f,
            limitDelta: _e8.visible ? r - _t7 : r + _t7,
            size: _e8.size
          };
        }
        if ("number" == typeof d) {
          var _e9 = _this8.viewItems[d],
            _t8 = Math.floor(_e9.viewMinimumSize / 2);
          n = {
            index: d,
            limitDelta: _e9.visible ? o + _t8 : o - _t8,
            size: _e9.size
          };
        }
        _this8.sashDragState = {
          start: e,
          current: e,
          index: i,
          sizes: t,
          minDelta: r,
          maxDelta: o,
          snapBefore: s,
          snapAfter: n
        };
      })(t);
    }
  }, {
    key: "onSashChange",
    value: function onSashChange(_ref4) {
      var e = _ref4.current;
      var _this$sashDragState = this.sashDragState,
        t = _this$sashDragState.index,
        i = _this$sashDragState.start,
        s = _this$sashDragState.sizes,
        n = _this$sashDragState.minDelta,
        r = _this$sashDragState.maxDelta,
        o = _this$sashDragState.snapBefore,
        a = _this$sashDragState.snapAfter;
      this.sashDragState.current = e;
      var h = e - i;
      this.resize(t, h, s, void 0, void 0, n, r, o, a), this.distributeEmptySpace(), this.layoutViews();
    }
  }, {
    key: "getSashPosition",
    value: function getSashPosition(e) {
      var t = 0;
      for (var _i8 = 0; _i8 < this.sashItems.length; _i8++) if (t += this.viewItems[_i8].size, this.sashItems[_i8].sash === e) return t;
      return 0;
    }
  }, {
    key: "resize",
    value: function resize(e, t) {
      var _this9 = this;
      var i = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this.viewItems.map(function (e) {
        return e.size;
      });
      var s = arguments.length > 3 ? arguments[3] : undefined;
      var n = arguments.length > 4 ? arguments[4] : undefined;
      var r = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : Number.NEGATIVE_INFINITY;
      var o = arguments.length > 6 && arguments[6] !== undefined ? arguments[6] : Number.POSITIVE_INFINITY;
      var a = arguments.length > 7 ? arguments[7] : undefined;
      var h = arguments.length > 8 ? arguments[8] : undefined;
      if (e < 0 || e >= this.viewItems.length) return 0;
      var l = Y(e, -1, -1),
        u = Y(e + 1, this.viewItems.length);
      if (n) {
        var _iterator7 = _createForOfIteratorHelper(n),
          _step7;
        try {
          for (_iterator7.s(); !(_step7 = _iterator7.n()).done;) {
            var _e0 = _step7.value;
            R(l, _e0), R(u, _e0);
          }
        } catch (err) {
          _iterator7.e(err);
        } finally {
          _iterator7.f();
        }
      }
      if (s) {
        var _iterator8 = _createForOfIteratorHelper(s),
          _step8;
        try {
          for (_iterator8.s(); !(_step8 = _iterator8.n()).done;) {
            var _e1 = _step8.value;
            k(l, _e1), k(u, _e1);
          }
        } catch (err) {
          _iterator8.e(err);
        } finally {
          _iterator8.f();
        }
      }
      var c = l.map(function (e) {
          return _this9.viewItems[e];
        }),
        m = l.map(function (e) {
          return i[e];
        }),
        f = u.map(function (e) {
          return _this9.viewItems[e];
        }),
        d = u.map(function (e) {
          return i[e];
        }),
        p = l.reduce(function (e, t) {
          return e + (_this9.viewItems[t].minimumSize - i[t]);
        }, 0),
        S = l.reduce(function (e, t) {
          return e + (_this9.viewItems[t].maximumSize - i[t]);
        }, 0),
        z = 0 === u.length ? Number.POSITIVE_INFINITY : u.reduce(function (e, t) {
          return e + (i[t] - _this9.viewItems[t].minimumSize);
        }, 0),
        w = 0 === u.length ? Number.NEGATIVE_INFINITY : u.reduce(function (e, t) {
          return e + (i[t] - _this9.viewItems[t].maximumSize);
        }, 0),
        g = Math.max(p, w, r),
        y = Math.min(z, S, o);
      var b = !1;
      if (a) {
        var _e10 = this.viewItems[a.index],
          _i9 = t >= a.limitDelta;
        b = _i9 !== _e10.visible, _e10.setVisible(_i9, a.size);
      }
      if (!b && h) {
        var _e11 = this.viewItems[h.index],
          _i0 = t < h.limitDelta;
        b = _i0 !== _e11.visible, _e11.setVisible(_i0, h.size);
      }
      if (b) return this.resize(e, t, i, s, n, r, o);
      for (var _e12 = 0, _i1 = t = v(t, g, y); _e12 < c.length; _e12++) {
        var _t9 = c[_e12],
          _s4 = v(m[_e12] + _i1, _t9.minimumSize, _t9.maximumSize);
        _i1 -= _s4 - m[_e12], _t9.size = _s4;
      }
      for (var _e13 = 0, _i10 = t; _e13 < f.length; _e13++) {
        var _t0 = f[_e13],
          _s5 = v(d[_e13] - _i10, _t0.minimumSize, _t0.maximumSize);
        _i10 += _s5 - d[_e13], _t0.size = _s5;
      }
      return t;
    }
  }, {
    key: "distributeEmptySpace",
    value: function distributeEmptySpace(e) {
      var _this0 = this;
      var t = this.viewItems.reduce(function (e, t) {
        return e + t.size;
      }, 0);
      var i = this.size - t;
      var s = Y(0, this.viewItems.length),
        n = [],
        r = s.filter(function (e) {
          return _this0.viewItems[e].priority === re.Low;
        }),
        o = s.filter(function (e) {
          return _this0.viewItems[e].priority === re.Normal;
        }),
        a = s.filter(function (e) {
          return _this0.viewItems[e].priority === re.High;
        });
      n.push.apply(n, _toConsumableArray(a).concat(_toConsumableArray(o), _toConsumableArray(r))), "number" == typeof e && k(n, e);
      for (var _e14 = 0; 0 !== i && _e14 < n.length; _e14++) {
        var _t1 = this.viewItems[n[_e14]],
          _s6 = v(_t1.size + i, _t1.minimumSize, _t1.maximumSize);
        i -= _s6 - _t1.size, _t1.size = _s6;
      }
    }
  }, {
    key: "layoutViews",
    value: function layoutViews() {
      var _this$onDidChange;
      this.contentSize = this.viewItems.reduce(function (e, t) {
        return e + t.size;
      }, 0);
      var e = 0;
      var _iterator9 = _createForOfIteratorHelper(this.viewItems),
        _step9;
      try {
        for (_iterator9.s(); !(_step9 = _iterator9.n()).done;) {
          var _t10 = _step9.value;
          _t10.layout(e), e += _t10.size;
        }
      } catch (err) {
        _iterator9.e(err);
      } finally {
        _iterator9.f();
      }
      (_this$onDidChange = this.onDidChange) !== null && _this$onDidChange !== void 0 && _this$onDidChange.call(this, this.viewItems.map(function (e) {
        return e.size;
      })), this.sashItems.forEach(function (e) {
        return e.sash.layout();
      }), this.updateSashEnablement();
    }
  }, {
    key: "saveProportions",
    value: function saveProportions() {
      var _this1 = this;
      this.proportionalLayout && this.contentSize > 0 && (this.proportions = this.viewItems.map(function (e) {
        return e.size / _this1.contentSize;
      }));
    }
  }, {
    key: "updateSashEnablement",
    value: function updateSashEnablement() {
      var e = !1;
      var t = this.viewItems.map(function (t) {
        return e = t.size - t.minimumSize > 0 || e;
      });
      e = !1;
      var i = this.viewItems.map(function (t) {
          return e = t.maximumSize - t.size > 0 || e;
        }),
        s = _toConsumableArray(this.viewItems).reverse();
      e = !1;
      var n = s.map(function (t) {
        return e = t.size - t.minimumSize > 0 || e;
      }).reverse();
      e = !1;
      var r = s.map(function (t) {
        return e = t.maximumSize - t.size > 0 || e;
      }).reverse();
      var o = 0;
      for (var _e15 = 0; _e15 < this.sashItems.length; _e15++) {
        var _s7 = this.sashItems[_e15].sash;
        o += this.viewItems[_e15].size;
        var _a = !(t[_e15] && r[_e15 + 1]),
          _h = !(i[_e15] && n[_e15 + 1]);
        if (_a && _h) {
          var _i11 = Y(_e15, -1, -1),
            _r2 = Y(_e15 + 1, this.viewItems.length),
            _a2 = this.findFirstSnapIndex(_i11),
            _h2 = this.findFirstSnapIndex(_r2),
            _l = "number" == typeof _a2 && !this.viewItems[_a2].visible,
            _u = "number" == typeof _h2 && !this.viewItems[_h2].visible;
          _l && n[_e15] && (o > 0 || this.startSnappingEnabled) ? _s7.state = Q.Minimum : _u && t[_e15] && (o < this.contentSize || this.endSnappingEnabled) ? _s7.state = Q.Maximum : _s7.state = Q.Disabled;
        } else _s7.state = _a && !_h ? Q.Minimum : !_a && _h ? Q.Maximum : Q.Enabled;
      }
    }
  }, {
    key: "findFirstSnapIndex",
    value: function findFirstSnapIndex(e) {
      var _iterator0 = _createForOfIteratorHelper(e),
        _step0;
      try {
        for (_iterator0.s(); !(_step0 = _iterator0.n()).done;) {
          var _t11 = _step0.value;
          var _e16 = this.viewItems[_t11];
          if (_e16.visible && _e16.snap) return _t11;
        }
      } catch (err) {
        _iterator0.e(err);
      } finally {
        _iterator0.f();
      }
      var _iterator1 = _createForOfIteratorHelper(e),
        _step1;
      try {
        for (_iterator1.s(); !(_step1 = _iterator1.n()).done;) {
          var _t12 = _step1.value;
          var _e17 = this.viewItems[_t12];
          if (_e17.visible && _e17.maximumSize - _e17.minimumSize > 0) return;
          if (!_e17.visible && _e17.snap) return _t12;
        }
      } catch (err) {
        _iterator1.e(err);
      } finally {
        _iterator1.f();
      }
    }
  }]);
}(H);
var ue = /*#__PURE__*/function () {
  function ue(e) {
    _classCallCheck(this, ue);
    this.size = void 0, this.size = e;
  }
  return _createClass(ue, [{
    key: "getPreferredSize",
    value: function getPreferredSize() {
      return this.size;
    }
  }]);
}();
var ce = /*#__PURE__*/function () {
  function ce(e, t) {
    _classCallCheck(this, ce);
    this.proportion = void 0, this.layoutService = void 0, this.proportion = e, this.layoutService = t;
  }
  return _createClass(ce, [{
    key: "getPreferredSize",
    value: function getPreferredSize() {
      return this.proportion * this.layoutService.getSize();
    }
  }]);
}();
var me = /*#__PURE__*/function () {
  function me() {
    _classCallCheck(this, me);
  }
  return _createClass(me, [{
    key: "getPreferredSize",
    value: function getPreferredSize() {}
  }]);
}();
var fe = /*#__PURE__*/function () {
  function fe(e, t) {
    var _t$priority;
    _classCallCheck(this, fe);
    if (this.minimumSize = 0, this.maximumSize = Number.POSITIVE_INFINITY, this.element = void 0, this.priority = void 0, this.snap = void 0, this.layoutService = void 0, this.layoutStrategy = void 0, this.layoutService = e, this.element = t.element, this.minimumSize = "number" == typeof t.minimumSize ? t.minimumSize : 30, this.maximumSize = "number" == typeof t.maximumSize ? t.maximumSize : Number.POSITIVE_INFINITY, "number" == typeof t.preferredSize) this.layoutStrategy = new ue(t.preferredSize);else if ("string" == typeof t.preferredSize) {
      var _e18 = t.preferredSize.trim();
      if (T(_e18, "%")) {
        var _t13 = Number(_e18.slice(0, -1)) / 100;
        this.layoutStrategy = new ce(_t13, this.layoutService);
      } else if (T(_e18, "px")) {
        var _t14 = Number(_e18.slice(0, -2));
        this.layoutStrategy = new ue(_t14);
      } else if ("number" == typeof Number.parseFloat(_e18)) {
        var _t15 = Number.parseFloat(_e18);
        this.layoutStrategy = new ue(_t15);
      } else this.layoutStrategy = new me();
    } else this.layoutStrategy = new me();
    this.priority = (_t$priority = t.priority) !== null && _t$priority !== void 0 ? _t$priority : re.Normal, this.snap = "boolean" == typeof t.snap && t.snap;
  }
  return _createClass(fe, [{
    key: "preferredSize",
    get: function get() {
      return this.layoutStrategy.getPreferredSize();
    },
    set: function set(e) {
      if ("number" == typeof e) this.layoutStrategy = new ue(e);else if ("string" == typeof e) {
        var _t16 = e.trim();
        if (T(_t16, "%")) {
          var _e19 = Number(_t16.slice(0, -1)) / 100;
          this.layoutStrategy = new ce(_e19, this.layoutService);
        } else if (T(_t16, "px")) {
          var _e20 = Number(_t16.slice(0, -2)) / 100;
          this.layoutStrategy = new ue(_e20);
        } else if ("number" == typeof Number.parseFloat(_t16)) {
          var _e21 = Number.parseFloat(_t16);
          this.layoutStrategy = new ue(_e21);
        } else this.layoutStrategy = new me();
      } else this.layoutStrategy = new me();
    }
  }, {
    key: "layout",
    value: function layout(e) {}
  }]);
}();
function de(e) {
  return void 0 !== e.minSize || void 0 !== e.maxSize || void 0 !== e.preferredSize || void 0 !== e.priority || void 0 !== e.visible;
}
var pe = e.forwardRef(function (_ref5, s) {
  var t = _ref5.className,
    i = _ref5.children;
  return e.createElement("div", {
    ref: s,
    className: l("split-view-view", x, t)
  }, i);
});
pe.displayName = "Allotment.Pane";
var ve = e.forwardRef(function (_ref6, g) {
  var t = _ref6.children,
    i = _ref6.className,
    s = _ref6.id,
    _ref6$maxSize = _ref6.maxSize,
    n = _ref6$maxSize === void 0 ? 1 / 0 : _ref6$maxSize,
    _ref6$minSize = _ref6.minSize,
    r = _ref6$minSize === void 0 ? 30 : _ref6$minSize,
    _ref6$proportionalLay = _ref6.proportionalLayout,
    o = _ref6$proportionalLay === void 0 ? !0 : _ref6$proportionalLay,
    _ref6$separator = _ref6.separator,
    a = _ref6$separator === void 0 ? !0 : _ref6$separator,
    h = _ref6.sizes,
    _ref6$defaultSizes = _ref6.defaultSizes,
    u = _ref6$defaultSizes === void 0 ? h : _ref6$defaultSizes,
    _ref6$snap = _ref6.snap,
    c = _ref6$snap === void 0 ? !1 : _ref6$snap,
    _ref6$vertical = _ref6.vertical,
    m = _ref6$vertical === void 0 ? !1 : _ref6$vertical,
    d = _ref6.onChange,
    p = _ref6.onReset,
    v = _ref6.onVisibleChange,
    S = _ref6.onDragStart,
    z = _ref6.onDragEnd;
  var b = e.useRef(null),
    x = e.useRef([]),
    N = e.useRef(new Map()),
    L = e.useRef(null),
    D = e.useRef(new Map()),
    O = e.useRef(new C()),
    T = e.useRef([]),
    _e$useState3 = e.useState(!1),
    _e$useState4 = _slicedToArray(_e$useState3, 2),
    A = _e$useState4[0],
    j = _e$useState4[1];
  "production" !== process.env.NODE_ENV && h && console.warn("Prop sizes is deprecated. Please use defaultSizes instead.");
  var F = e.useMemo(function () {
      return e.Children.toArray(t).filter(e.isValidElement);
    }, [t]),
    H = e.useCallback(function (e) {
      var _T$current, _L$current;
      var t = (_T$current = T.current) === null || _T$current === void 0 ? void 0 : _T$current[e];
      return "number" == typeof (t === null || t === void 0 ? void 0 : t.preferredSize) && ((_L$current = L.current) !== null && _L$current !== void 0 && _L$current.resizeView(e, Math.round(t.preferredSize)), !0);
    }, []);
  return e.useImperativeHandle(g, function () {
    return {
      reset: function reset() {
        if (p) p();else {
          var _L$current2;
          (_L$current2 = L.current) === null || _L$current2 === void 0 || _L$current2.distributeViewSizes();
          for (var _e22 = 0; _e22 < T.current.length; _e22++) H(_e22);
        }
      },
      resize: function resize(e) {
        var _L$current3;
        (_L$current3 = L.current) === null || _L$current3 === void 0 || _L$current3.resizeViews(e);
      }
    };
  }), P(function () {
    var e = !0;
    u && D.current.size !== u.length && (e = !1, console.warn("Expected ".concat(u.length, " children based on defaultSizes but found ").concat(D.current.size))), e && u && (x.current = F.map(function (e) {
      return e.key;
    }));
    var t = _objectSpread({
      orientation: m ? q.Vertical : q.Horizontal,
      proportionalLayout: o
    }, e && u && {
      descriptor: {
        size: u.reduce(function (e, t) {
          return e + t;
        }, 0),
        views: u.map(function (e, t) {
          var _i$minSize, _i$maxSize, _i$priority, _i$snap;
          var i = N.current.get(x.current[t]),
            s = new fe(O.current, _objectSpread(_objectSpread({
              element: document.createElement("div"),
              minimumSize: (_i$minSize = i === null || i === void 0 ? void 0 : i.minSize) !== null && _i$minSize !== void 0 ? _i$minSize : r,
              maximumSize: (_i$maxSize = i === null || i === void 0 ? void 0 : i.maxSize) !== null && _i$maxSize !== void 0 ? _i$maxSize : n,
              priority: (_i$priority = i === null || i === void 0 ? void 0 : i.priority) !== null && _i$priority !== void 0 ? _i$priority : re.Normal
            }, (i === null || i === void 0 ? void 0 : i.preferredSize) && {
              preferredSize: i === null || i === void 0 ? void 0 : i.preferredSize
            }), {}, {
              snap: (_i$snap = i === null || i === void 0 ? void 0 : i.snap) !== null && _i$snap !== void 0 ? _i$snap : c
            }));
          return T.current.push(s), {
            container: _toConsumableArray(D.current.values())[t],
            size: e,
            view: s
          };
        })
      }
    });
    L.current = new le(b.current, t, d, S, z), L.current.on("sashDragStart", function () {
      var _b$current;
      (_b$current = b.current) === null || _b$current === void 0 || _b$current.classList.add("split-view-sash-dragging");
    }), L.current.on("sashDragEnd", function () {
      var _b$current2;
      (_b$current2 = b.current) === null || _b$current2 === void 0 || _b$current2.classList.remove("split-view-sash-dragging");
    }), L.current.on("sashchange", function (e) {
      if (v && L.current) {
        var _e23 = F.map(function (e) {
          return e.key;
        });
        for (var _t17 = 0; _t17 < _e23.length; _t17++) {
          var _i12 = N.current.get(_e23[_t17]);
          void 0 !== (_i12 === null || _i12 === void 0 ? void 0 : _i12.visible) && _i12.visible !== L.current.isViewVisible(_t17) && v(_t17, L.current.isViewVisible(_t17));
        }
      }
    }), L.current.on("sashreset", function (e) {
      if (p) p();else {
        var _L$current4;
        if (H(e)) return;
        if (H(e + 1)) return;
        (_L$current4 = L.current) === null || _L$current4 === void 0 || _L$current4.distributeViewSizes();
      }
    });
    var i = L.current;
    return function () {
      i.dispose();
    };
  }, []), P(function () {
    if (A) {
      var _e24 = F.map(function (e) {
          return e.key;
        }),
        _t18 = _toConsumableArray(x.current),
        _i13 = _e24.filter(function (e) {
          return !x.current.includes(e);
        }),
        _s8 = _e24.filter(function (e) {
          return x.current.includes(e);
        }),
        _o = x.current.map(function (t) {
          return !_e24.includes(t);
        });
      for (var _e25 = _o.length - 1; _e25 >= 0; _e25--) {
        var _L$current5;
        _o[_e25] && ((_L$current5 = L.current) !== null && _L$current5 !== void 0 && _L$current5.removeView(_e25), _t18.splice(_e25, 1), T.current.splice(_e25, 1));
      }
      var _iterator10 = _createForOfIteratorHelper(_i13),
        _step10;
      try {
        var _loop3 = function _loop3() {
          var _i$minSize2, _i$maxSize2, _i$priority2, _i$snap2, _L$current9;
          var s = _step10.value;
          var i = N.current.get(s),
            o = new fe(O.current, _objectSpread(_objectSpread({
              element: document.createElement("div"),
              minimumSize: (_i$minSize2 = i === null || i === void 0 ? void 0 : i.minSize) !== null && _i$minSize2 !== void 0 ? _i$minSize2 : r,
              maximumSize: (_i$maxSize2 = i === null || i === void 0 ? void 0 : i.maxSize) !== null && _i$maxSize2 !== void 0 ? _i$maxSize2 : n,
              priority: (_i$priority2 = i === null || i === void 0 ? void 0 : i.priority) !== null && _i$priority2 !== void 0 ? _i$priority2 : re.Normal
            }, (i === null || i === void 0 ? void 0 : i.preferredSize) && {
              preferredSize: i === null || i === void 0 ? void 0 : i.preferredSize
            }), {}, {
              snap: (_i$snap2 = i === null || i === void 0 ? void 0 : i.snap) !== null && _i$snap2 !== void 0 ? _i$snap2 : c
            }));
          (_L$current9 = L.current) !== null && _L$current9 !== void 0 && _L$current9.addView(D.current.get(s), o, se.Distribute, _e24.findIndex(function (e) {
            return e === s;
          })), _t18.splice(_e24.findIndex(function (e) {
            return e === s;
          }), 0, s), T.current.splice(_e24.findIndex(function (e) {
            return e === s;
          }), 0, o);
        };
        for (_iterator10.s(); !(_step10 = _iterator10.n()).done;) {
          _loop3();
        }
      } catch (err) {
        _iterator10.e(err);
      } finally {
        _iterator10.f();
      }
      for (; !f(_e24, _t18);) {
        var _iterator11 = _createForOfIteratorHelper(_e24.entries()),
          _step11;
        try {
          var _loop = function _loop() {
            var _step11$value = _slicedToArray(_step11.value, 2),
              i = _step11$value[0],
              s = _step11$value[1];
            var e = _t18.findIndex(function (e) {
              return e === s;
            });
            if (e !== i) {
              var _L$current6;
              (_L$current6 = L.current) === null || _L$current6 === void 0 || _L$current6.moveView(D.current.get(s), e, i);
              var _n3 = _t18[e];
              _t18.splice(e, 1), _t18.splice(i, 0, _n3);
              return 1; // break
            }
          };
          for (_iterator11.s(); !(_step11 = _iterator11.n()).done;) {
            if (_loop()) break;
          }
        } catch (err) {
          _iterator11.e(err);
        } finally {
          _iterator11.f();
        }
      }
      var _iterator12 = _createForOfIteratorHelper(_i13),
        _step12;
      try {
        var _loop4 = function _loop4() {
          var _L$current0;
          var t = _step12.value;
          var i = _e24.findIndex(function (e) {
              return e === t;
            }),
            s = T.current[i].preferredSize;
          void 0 !== s && ((_L$current0 = L.current) === null || _L$current0 === void 0 ? void 0 : _L$current0.resizeView(i, s));
        };
        for (_iterator12.s(); !(_step12 = _iterator12.n()).done;) {
          _loop4();
        }
      } catch (err) {
        _iterator12.e(err);
      } finally {
        _iterator12.f();
      }
      var _loop2 = function _loop2() {
        var _L$current7, _L$current8;
        var t = _arr[_i14];
        var i = N.current.get(t),
          s = _e24.findIndex(function (e) {
            return e === t;
          });
        i && de(i) && void 0 !== i.visible && ((_L$current7 = L.current) === null || _L$current7 === void 0 ? void 0 : _L$current7.isViewVisible(s)) !== i.visible && ((_L$current8 = L.current) === null || _L$current8 === void 0 ? void 0 : _L$current8.setViewVisible(s, i.visible));
      };
      for (var _i14 = 0, _arr = [].concat(_toConsumableArray(_i13), _toConsumableArray(_s8)); _i14 < _arr.length; _i14++) {
        _loop2();
      }
      var _iterator13 = _createForOfIteratorHelper(_s8),
        _step13;
      try {
        var _loop5 = function _loop5() {
          var t = _step13.value;
          var i = N.current.get(t),
            s = _e24.findIndex(function (e) {
              return e === t;
            });
          if (i && de(i)) {
            var _L$current1;
            void 0 !== i.preferredSize && T.current[s].preferredSize !== i.preferredSize && (T.current[s].preferredSize = i.preferredSize);
            var _e26 = !1;
            void 0 !== i.minSize && T.current[s].minimumSize !== i.minSize && (T.current[s].minimumSize = i.minSize, _e26 = !0), void 0 !== i.maxSize && T.current[s].maximumSize !== i.maxSize && (T.current[s].maximumSize = i.maxSize, _e26 = !0), _e26 && ((_L$current1 = L.current) === null || _L$current1 === void 0 ? void 0 : _L$current1.layout());
          }
        };
        for (_iterator13.s(); !(_step13 = _iterator13.n()).done;) {
          _loop5();
        }
      } catch (err) {
        _iterator13.e(err);
      } finally {
        _iterator13.f();
      }
      (_i13.length > 0 || _o.length > 0) && (x.current = _e24);
    }
  }, [F, A, n, r, c]), e.useEffect(function () {
    L.current && (L.current.onDidChange = d);
  }, [d]), e.useEffect(function () {
    L.current && (L.current.onDidDragStart = S);
  }, [S]), e.useEffect(function () {
    L.current && (L.current.onDidDragEnd = z);
  }, [z]), w({
    ref: b,
    onResize: function onResize(_ref7) {
      var _L$current10;
      var e = _ref7.width,
        t = _ref7.height;
      e && t && ((_L$current10 = L.current) !== null && _L$current10 !== void 0 && _L$current10.layout(m ? t : e), O.current.setSize(m ? t : e), j(!0));
    }
  }), e.useEffect(function () {
    M && Se(20);
  }, []), e.createElement("div", {
    ref: b,
    className: l("split-view", m ? "split-view-vertical" : "split-view-horizontal", {
      "split-view-separator-border": a
    }, y, m ? _ : E, _defineProperty({}, V, a), i),
    id: s
  }, e.createElement("div", {
    className: l("split-view-container", I)
  }, e.Children.toArray(t).map(function (t) {
    if (!e.isValidElement(t)) return null;
    var i = t.key;
    return "Allotment.Pane" === t.type.displayName ? (N.current.set(i, t.props), e.cloneElement(t, {
      key: i,
      ref: function ref(e) {
        var s = t.ref;
        s && (s.current = e), e ? D.current.set(i, e) : D.current["delete"](i);
      }
    })) : e.createElement(pe, {
      key: i,
      ref: function ref(e) {
        e ? D.current.set(i, e) : D.current["delete"](i);
      }
    }, t);
  })));
});
function Se(e) {
  var t = v(e, 4, 20),
    i = v(e, 1, 8);
  document.documentElement.style.setProperty("--sash-size", t + "px"), document.documentElement.style.setProperty("--sash-hover-size", i + "px"), function (e) {
    ee = e, te.emit("onDidChangeGlobalSize", e);
  }(t);
}
ve.displayName = "Allotment";
var ze = Object.assign(ve, {
  Pane: pe
});
exports.Allotment = ze, exports.LayoutPriority = re, exports.setSashSize = Se;
