// allotment/dist/module.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import e, { useState as t, useRef as i, useEffect as s, useCallback as n, useLayoutEffect as r, forwardRef as o, useMemo as a, useImperativeHandle as h } from "react";
var l = "undefined" != typeof globalThis ? globalThis : "undefined" != typeof window ? window : "undefined" != typeof global ? global : "undefined" != typeof self ? self : {};
function u(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var c,
  m = {
    exports: {}
  };
/*!
	Copyright (c) 2018 Jed Watson.
	Licensed under the MIT License (MIT), see
	http://jedwatson.github.io/classnames
*/
var d,
  f,
  p,
  v = (c || (c = 1, d = m, function () {
    var e = {}.hasOwnProperty;
    function t() {
      for (var e = "", t = 0; t < arguments.length; t++) {
        var n = arguments[t];
        n && (e = s(e, i(n)));
      }
      return e;
    }
    function i(i) {
      if ("string" == typeof i || "number" == typeof i) return i;
      if ("object" != typeof i) return "";
      if (Array.isArray(i)) return t.apply(null, i);
      if (i.toString !== Object.prototype.toString && !i.toString.toString().includes("[native code]")) return i.toString();
      var n = "";
      for (var r in i) e.call(i, r) && i[r] && (n = s(n, r));
      return n;
    }
    function s(e, t) {
      return t ? e ? e + " " + t : e + t : e;
    }
    d.exports ? (t.default = t, d.exports = t) : window.classNames = t;
  }()), m.exports),
  S = u(v);
var z,
  w,
  g = (p || (p = 1, f = function e(t, i) {
    if (t === i) return !0;
    if (t && i && "object" == typeof t && "object" == typeof i) {
      if (t.constructor !== i.constructor) return !1;
      var s, n, r;
      if (Array.isArray(t)) {
        if ((s = t.length) != i.length) return !1;
        for (n = s; 0 !== n--;) if (!e(t[n], i[n])) return !1;
        return !0;
      }
      if (t instanceof Map && i instanceof Map) {
        if (t.size !== i.size) return !1;
        for (n of t.entries()) if (!i.has(n[0])) return !1;
        for (n of t.entries()) if (!e(n[1], i.get(n[0]))) return !1;
        return !0;
      }
      if (t instanceof Set && i instanceof Set) {
        if (t.size !== i.size) return !1;
        for (n of t.entries()) if (!i.has(n[0])) return !1;
        return !0;
      }
      if (ArrayBuffer.isView(t) && ArrayBuffer.isView(i)) {
        if ((s = t.length) != i.length) return !1;
        for (n = s; 0 !== n--;) if (t[n] !== i[n]) return !1;
        return !0;
      }
      if (t.constructor === RegExp) return t.source === i.source && t.flags === i.flags;
      if (t.valueOf !== Object.prototype.valueOf) return t.valueOf() === i.valueOf();
      if (t.toString !== Object.prototype.toString) return t.toString() === i.toString();
      if ((s = (r = Object.keys(t)).length) !== Object.keys(i).length) return !1;
      for (n = s; 0 !== n--;) if (!Object.prototype.hasOwnProperty.call(i, r[n])) return !1;
      for (n = s; 0 !== n--;) {
        var o = r[n];
        if (!e(t[o], i[o])) return !1;
      }
      return !0;
    }
    return t != t && i != i;
  }), f),
  y = u(g);
var b,
  I,
  x = u(function () {
    if (w) return z;
    w = 1;
    var e = /^\s+|\s+$/g,
      t = /^[-+]0x[0-9a-f]+$/i,
      i = /^0b[01]+$/i,
      s = /^0o[0-7]+$/i,
      n = parseInt,
      r = Object.prototype.toString;
    function o(e) {
      var t = typeof e;
      return !!e && ("object" == t || "function" == t);
    }
    function a(a) {
      if ("number" == typeof a) return a;
      if (function (e) {
        return "symbol" == typeof e || function (e) {
          return !!e && "object" == typeof e;
        }(e) && "[object Symbol]" == r.call(e);
      }(a)) return NaN;
      if (o(a)) {
        var h = "function" == typeof a.valueOf ? a.valueOf() : a;
        a = o(h) ? h + "" : h;
      }
      if ("string" != typeof a) return 0 === a ? a : +a;
      a = a.replace(e, "");
      var l = i.test(a);
      return l || s.test(a) ? n(a.slice(2), l ? 2 : 8) : t.test(a) ? NaN : +a;
    }
    return z = function z(e, t, i) {
      return void 0 === i && (i = t, t = void 0), void 0 !== i && (i = (i = a(i)) == i ? i : 0), void 0 !== t && (t = (t = a(t)) == t ? t : 0), function (e, t, i) {
        return e == e && (void 0 !== i && (e = e <= i ? e : i), void 0 !== t && (e = e >= t ? e : t)), e;
      }(a(e), t, i);
    };
  }());
var _ = u(function () {
  if (I) return b;
  I = 1;
  var e = /^\s+|\s+$/g,
    t = /^[-+]0x[0-9a-f]+$/i,
    i = /^0b[01]+$/i,
    s = /^0o[0-7]+$/i,
    n = parseInt,
    r = "object" == typeof l && l && l.Object === Object && l,
    o = "object" == typeof self && self && self.Object === Object && self,
    a = r || o || Function("return this")(),
    h = Object.prototype.toString,
    u = Math.max,
    c = Math.min,
    m = function m() {
      return a.Date.now();
    };
  function d(e) {
    var t = typeof e;
    return !!e && ("object" == t || "function" == t);
  }
  function f(r) {
    if ("number" == typeof r) return r;
    if (function (e) {
      return "symbol" == typeof e || function (e) {
        return !!e && "object" == typeof e;
      }(e) && "[object Symbol]" == h.call(e);
    }(r)) return NaN;
    if (d(r)) {
      var o = "function" == typeof r.valueOf ? r.valueOf() : r;
      r = d(o) ? o + "" : o;
    }
    if ("string" != typeof r) return 0 === r ? r : +r;
    r = r.replace(e, "");
    var a = i.test(r);
    return a || s.test(r) ? n(r.slice(2), a ? 2 : 8) : t.test(r) ? NaN : +r;
  }
  return b = function b(e, t, i) {
    var s,
      n,
      r,
      o,
      a,
      h,
      l = 0,
      p = !1,
      v = !1,
      S = !0;
    if ("function" != typeof e) throw new TypeError("Expected a function");
    function z(t) {
      var i = s,
        r = n;
      return s = n = void 0, l = t, o = e.apply(r, i);
    }
    function w(e) {
      var i = e - h;
      return void 0 === h || i >= t || i < 0 || v && e - l >= r;
    }
    function g() {
      var e = m();
      if (w(e)) return y(e);
      a = setTimeout(g, function (e) {
        var i = t - (e - h);
        return v ? c(i, r - (e - l)) : i;
      }(e));
    }
    function y(e) {
      return a = void 0, S && s ? z(e) : (s = n = void 0, o);
    }
    function b() {
      var e = m(),
        i = w(e);
      if (s = arguments, n = this, h = e, i) {
        if (void 0 === a) return function (e) {
          return l = e, a = setTimeout(g, t), p ? z(e) : o;
        }(h);
        if (v) return a = setTimeout(g, t), z(h);
      }
      return void 0 === a && (a = setTimeout(g, t)), o;
    }
    return t = f(t) || 0, d(i) && (p = !!i.leading, r = (v = "maxWait" in i) ? u(f(i.maxWait) || 0, t) : r, S = "trailing" in i ? !!i.trailing : S), b.cancel = function () {
      void 0 !== a && clearTimeout(a), l = 0, s = h = n = a = void 0;
    }, b.flush = function () {
      return void 0 === a ? o : y(m());
    }, b;
  };
}());
var V = {
  width: void 0,
  height: void 0
};
function E(e) {
  var {
      ref: r,
      box: o = "content-box"
    } = e,
    [{
      width: a,
      height: h
    }, l] = t(V),
    u = function () {
      var e = i(!1);
      return s(() => (e.current = !0, () => {
        e.current = !1;
      }), []), n(() => e.current, []);
    }(),
    c = i(_extends({}, V)),
    m = i(void 0);
  return m.current = e.onResize, s(() => {
    if (!r.current) return;
    if ("undefined" == typeof window || !("ResizeObserver" in window)) return;
    var e = new ResizeObserver(_ref => {
      var [e] = _ref;
      var t = "border-box" === o ? "borderBoxSize" : "device-pixel-content-box" === o ? "devicePixelContentBoxSize" : "contentBoxSize",
        i = N(e, t, "inlineSize"),
        s = N(e, t, "blockSize");
      if (c.current.width !== i || c.current.height !== s) {
        var _e2 = {
          width: i,
          height: s
        };
        c.current.width = i, c.current.height = s, m.current ? m.current(_e2) : u() && l(_e2);
      }
    });
    return e.observe(r.current, {
      box: o
    }), () => {
      e.disconnect();
    };
  }, [o, r, u]), {
    width: a,
    height: h
  };
}
function N(e, t, i) {
  return e[t] ? Array.isArray(e[t]) ? e[t][0][i] : e[t][i] : "contentBoxSize" === t ? e.contentRect["inlineSize" === i ? "width" : "height"] : void 0;
}
var L = "allotment-module_splitView__L-yRc",
  D = "allotment-module_sashContainer__fzwJF",
  O = "allotment-module_splitViewContainer__rQnVa",
  M = "allotment-module_splitViewView__MGZ6O",
  P = "allotment-module_vertical__WSwwa",
  T = "allotment-module_horizontal__7doS8",
  C = "allotment-module_separatorBorder__x-rDS";
var A,
  j = !1,
  F = !1;
"object" == typeof navigator && (A = navigator.userAgent, F = A.indexOf("Macintosh") >= 0, j = (A.indexOf("Macintosh") >= 0 || A.indexOf("iPad") >= 0 || A.indexOf("iPhone") >= 0) && !!navigator.maxTouchPoints && navigator.maxTouchPoints > 0);
var H = j,
  Y = F,
  k = "undefined" != typeof window && void 0 !== window.document && void 0 !== window.document.createElement ? r : s;
class $ {
  constructor() {
    this._size = void 0;
  }
  getSize() {
    return this._size;
  }
  setSize(e) {
    this._size = e;
  }
}
function B(e, t) {
  var i = e.length,
    s = i - t.length;
  return s >= 0 && e.slice(s, i) === t;
}
var R,
  W = {
    exports: {}
  };
var X = (R || (R = 1, function (e) {
    var t = Object.prototype.hasOwnProperty,
      i = "~";
    function s() {}
    function n(e, t, i) {
      this.fn = e, this.context = t, this.once = i || !1;
    }
    function r(e, t, s, r, o) {
      if ("function" != typeof s) throw new TypeError("The listener must be a function");
      var a = new n(s, r || e, o),
        h = i ? i + t : t;
      return e._events[h] ? e._events[h].fn ? e._events[h] = [e._events[h], a] : e._events[h].push(a) : (e._events[h] = a, e._eventsCount++), e;
    }
    function o(e, t) {
      0 === --e._eventsCount ? e._events = new s() : delete e._events[t];
    }
    function a() {
      this._events = new s(), this._eventsCount = 0;
    }
    Object.create && (s.prototype = Object.create(null), new s().__proto__ || (i = !1)), a.prototype.eventNames = function () {
      var e,
        s,
        n = [];
      if (0 === this._eventsCount) return n;
      for (s in e = this._events) t.call(e, s) && n.push(i ? s.slice(1) : s);
      return Object.getOwnPropertySymbols ? n.concat(Object.getOwnPropertySymbols(e)) : n;
    }, a.prototype.listeners = function (e) {
      var t = i ? i + e : e,
        s = this._events[t];
      if (!s) return [];
      if (s.fn) return [s.fn];
      for (var n = 0, r = s.length, o = new Array(r); n < r; n++) o[n] = s[n].fn;
      return o;
    }, a.prototype.listenerCount = function (e) {
      var t = i ? i + e : e,
        s = this._events[t];
      return s ? s.fn ? 1 : s.length : 0;
    }, a.prototype.emit = function (e, t, s, n, r, o) {
      var a = i ? i + e : e;
      if (!this._events[a]) return !1;
      var h,
        l,
        u = this._events[a],
        c = arguments.length;
      if (u.fn) {
        switch (u.once && this.removeListener(e, u.fn, void 0, !0), c) {
          case 1:
            return u.fn.call(u.context), !0;
          case 2:
            return u.fn.call(u.context, t), !0;
          case 3:
            return u.fn.call(u.context, t, s), !0;
          case 4:
            return u.fn.call(u.context, t, s, n), !0;
          case 5:
            return u.fn.call(u.context, t, s, n, r), !0;
          case 6:
            return u.fn.call(u.context, t, s, n, r, o), !0;
        }
        for (l = 1, h = new Array(c - 1); l < c; l++) h[l - 1] = arguments[l];
        u.fn.apply(u.context, h);
      } else {
        var m,
          d = u.length;
        for (l = 0; l < d; l++) switch (u[l].once && this.removeListener(e, u[l].fn, void 0, !0), c) {
          case 1:
            u[l].fn.call(u[l].context);
            break;
          case 2:
            u[l].fn.call(u[l].context, t);
            break;
          case 3:
            u[l].fn.call(u[l].context, t, s);
            break;
          case 4:
            u[l].fn.call(u[l].context, t, s, n);
            break;
          default:
            if (!h) for (m = 1, h = new Array(c - 1); m < c; m++) h[m - 1] = arguments[m];
            u[l].fn.apply(u[l].context, h);
        }
      }
      return !0;
    }, a.prototype.on = function (e, t, i) {
      return r(this, e, t, i, !1);
    }, a.prototype.once = function (e, t, i) {
      return r(this, e, t, i, !0);
    }, a.prototype.removeListener = function (e, t, s, n) {
      var r = i ? i + e : e;
      if (!this._events[r]) return this;
      if (!t) return o(this, r), this;
      var a = this._events[r];
      if (a.fn) a.fn !== t || n && !a.once || s && a.context !== s || o(this, r);else {
        for (var h = 0, l = [], u = a.length; h < u; h++) (a[h].fn !== t || n && !a[h].once || s && a[h].context !== s) && l.push(a[h]);
        l.length ? this._events[r] = 1 === l.length ? l[0] : l : o(this, r);
      }
      return this;
    }, a.prototype.removeAllListeners = function (e) {
      var t;
      return e ? (t = i ? i + e : e, this._events[t] && o(this, t)) : (this._events = new s(), this._eventsCount = 0), this;
    }, a.prototype.off = a.prototype.removeListener, a.prototype.addListener = a.prototype.on, a.prefixed = i, a.EventEmitter = a, e.exports = a;
  }(W)), W.exports),
  G = u(X);
function J(e, t) {
  var i = e.indexOf(t);
  i > -1 && (e.splice(i, 1), e.unshift(t));
}
function U(e, t) {
  var i = e.indexOf(t);
  i > -1 && (e.splice(i, 1), e.push(t));
}
function K(e, t, i) {
  if (i === void 0) {
    i = 1;
  }
  var s = Math.max(0, Math.ceil((t - e) / i)),
    n = new Array(s);
  var r = -1;
  for (; ++r < s;) n[r] = e + r * i;
  return n;
}
var Z = "sash-module_sash__K-9lB",
  Q = "sash-module_disabled__Hm-wx",
  q = "sash-module_mac__Jf6OJ",
  ee = "sash-module_vertical__pB-rs",
  te = "sash-module_minimum__-UKxp",
  ie = "sash-module_maximum__TCWxD",
  se = "sash-module_horizontal__kFbiw",
  ne = "sash-module_hover__80W6I",
  re = "sash-module_active__bJspD";
var oe = function (e) {
    return e.Vertical = "VERTICAL", e.Horizontal = "HORIZONTAL", e;
  }({}),
  ae = function (e) {
    return e.Disabled = "DISABLED", e.Minimum = "MINIMUM", e.Maximum = "MAXIMUM", e.Enabled = "ENABLED", e;
  }({}),
  he = H ? 20 : 8;
var le = new G();
class ue extends G {
  get state() {
    return this._state;
  }
  set state(e) {
    this._state !== e && (this.el.classList.toggle(Q, e === ae.Disabled), this.el.classList.toggle("sash-disabled", e === ae.Disabled), this.el.classList.toggle(te, e === ae.Minimum), this.el.classList.toggle("sash-minimum", e === ae.Minimum), this.el.classList.toggle(ie, e === ae.Maximum), this.el.classList.toggle("sash-maximum", e === ae.Maximum), this._state = e, this.emit("enablementChange", e));
  }
  constructor(e, t, i) {
    var _i$orientation;
    super(), this.el = void 0, this.layoutProvider = void 0, this.orientation = void 0, this.size = void 0, this.hoverDelay = 300, this.hoverDelayer = _(e => e.classList.add("sash-hover", ne), this.hoverDelay), this._state = ae.Enabled, this.onPointerStart = e => {
      var t = e.pageX,
        i = e.pageY,
        s = {
          startX: t,
          currentX: t,
          startY: i,
          currentY: i
        };
      this.el.classList.add("sash-active", re), this.emit("start", s), this.el.setPointerCapture(e.pointerId);
      var n = e => {
          e.preventDefault();
          var s = {
            startX: t,
            currentX: e.pageX,
            startY: i,
            currentY: e.pageY
          };
          this.emit("change", s);
        },
        r = e => {
          e.preventDefault(), this.el.classList.remove("sash-active", re), this.hoverDelayer.cancel(), this.emit("end"), this.el.releasePointerCapture(e.pointerId), window.removeEventListener("pointermove", n), window.removeEventListener("pointerup", r);
        };
      window.addEventListener("pointermove", n), window.addEventListener("pointerup", r);
    }, this.onPointerDoublePress = () => {
      this.emit("reset");
    }, this.onMouseEnter = () => {
      this.el.classList.contains(re) ? (this.hoverDelayer.cancel(), this.el.classList.add("sash-hover", ne)) : this.hoverDelayer(this.el);
    }, this.onMouseLeave = () => {
      this.hoverDelayer.cancel(), this.el.classList.remove("sash-hover", ne);
    }, this.el = document.createElement("div"), this.el.classList.add("sash", Z), this.el.dataset.testid = "sash", e.append(this.el), Y && this.el.classList.add("sash-mac", q), this.el.addEventListener("pointerdown", this.onPointerStart), this.el.addEventListener("dblclick", this.onPointerDoublePress), this.el.addEventListener("mouseenter", this.onMouseEnter), this.el.addEventListener("mouseleave", this.onMouseLeave), "number" == typeof i.size ? (this.size = i.size, i.orientation === oe.Vertical ? this.el.style.width = this.size + "px" : this.el.style.height = this.size + "px") : (this.size = he, le.on("onDidChangeGlobalSize", e => {
      this.size = e, this.layout();
    })), this.layoutProvider = t, this.orientation = (_i$orientation = i.orientation) != null ? _i$orientation : oe.Vertical, this.orientation === oe.Horizontal ? (this.el.classList.add("sash-horizontal", se), this.el.classList.remove("sash-vertical", ee)) : (this.el.classList.remove("sash-horizontal", se), this.el.classList.add("sash-vertical", ee)), this.layout();
  }
  layout() {
    if (this.orientation === oe.Vertical) {
      var _e3 = this.layoutProvider;
      this.el.style.left = _e3.getVerticalSashLeft(this) - this.size / 2 + "px", _e3.getVerticalSashTop && (this.el.style.top = _e3.getVerticalSashTop(this) + "px"), _e3.getVerticalSashHeight && (this.el.style.height = _e3.getVerticalSashHeight(this) + "px");
    } else {
      var _e4 = this.layoutProvider;
      this.el.style.top = _e4.getHorizontalSashTop(this) - this.size / 2 + "px", _e4.getHorizontalSashLeft && (this.el.style.left = _e4.getHorizontalSashLeft(this) + "px"), _e4.getHorizontalSashWidth && (this.el.style.width = _e4.getHorizontalSashWidth(this) + "px");
    }
  }
  dispose() {
    this.el.removeEventListener("pointerdown", this.onPointerStart), this.el.removeEventListener("dblclick", this.onPointerDoublePress), this.el.removeEventListener("mouseenter", this.onMouseEnter), this.el.removeEventListener("mouseleave", () => this.onMouseLeave), this.el.remove();
  }
}
var ce;
var me;
(me = ce || (ce = {})).Distribute = {
  type: "distribute"
}, me.Split = function (e) {
  return {
    type: "split",
    index: e
  };
}, me.Invisible = function (e) {
  return {
    type: "invisible",
    cachedVisibleSize: e
  };
};
var de = function (e) {
  return e.Normal = "NORMAL", e.Low = "LOW", e.High = "HIGH", e;
}({});
class fe {
  constructor(e, t, i) {
    this.container = void 0, this.view = void 0, this._size = void 0, this._cachedVisibleSize = void 0, this.container = e, this.view = t, this.container.classList.add("split-view-view", M), this.container.dataset.testid = "split-view-view", "number" == typeof i ? (this._size = i, this._cachedVisibleSize = void 0, e.classList.add("split-view-view-visible")) : (this._size = 0, this._cachedVisibleSize = i.cachedVisibleSize);
  }
  set size(e) {
    this._size = e;
  }
  get size() {
    return this._size;
  }
  get priority() {
    return this.view.priority;
  }
  get snap() {
    return !!this.view.snap;
  }
  get cachedVisibleSize() {
    return this._cachedVisibleSize;
  }
  get visible() {
    return void 0 === this._cachedVisibleSize;
  }
  setVisible(e, t) {
    e !== this.visible && (e ? (this.size = x(this._cachedVisibleSize, this.viewMinimumSize, this.viewMaximumSize), this._cachedVisibleSize = void 0) : (this._cachedVisibleSize = "number" == typeof t ? t : this.size, this.size = 0), this.container.classList.toggle("split-view-view-visible", e), this.view.setVisible && this.view.setVisible(e));
  }
  get minimumSize() {
    return this.visible ? this.view.minimumSize : 0;
  }
  get viewMinimumSize() {
    return this.view.minimumSize;
  }
  get maximumSize() {
    return this.visible ? this.view.maximumSize : 0;
  }
  get viewMaximumSize() {
    return this.view.maximumSize;
  }
  set enabled(e) {
    this.container.style.pointerEvents = e ? "" : "none";
  }
  layout(e) {
    this.layoutContainer(e), this.view.layout(this.size, e);
  }
}
class pe extends fe {
  layoutContainer(e) {
    this.container.style.left = e + "px", this.container.style.width = this.size + "px";
  }
}
class ve extends fe {
  layoutContainer(e) {
    this.container.style.top = e + "px", this.container.style.height = this.size + "px";
  }
}
class Se extends G {
  get startSnappingEnabled() {
    return this._startSnappingEnabled;
  }
  set startSnappingEnabled(e) {
    this._startSnappingEnabled !== e && (this._startSnappingEnabled = e, this.updateSashEnablement());
  }
  get endSnappingEnabled() {
    return this._endSnappingEnabled;
  }
  set endSnappingEnabled(e) {
    this._endSnappingEnabled !== e && (this._endSnappingEnabled = e, this.updateSashEnablement());
  }
  constructor(e, t, i, s, n) {
    var _t$orientation, _t$proportionalLayout;
    if (t === void 0) {
      t = {};
    }
    if (super(), this.onDidChange = void 0, this.onDidDragStart = void 0, this.onDidDragEnd = void 0, this.orientation = void 0, this.sashContainer = void 0, this.size = 0, this.contentSize = 0, this.proportions = void 0, this.viewItems = [], this.sashItems = [], this.sashDragState = void 0, this.proportionalLayout = void 0, this.getSashOrthogonalSize = void 0, this._startSnappingEnabled = !0, this._endSnappingEnabled = !0, this.onSashEnd = e => {
      this.emit("sashchange", e), this.saveProportions();
      for (var _e5 of this.viewItems) _e5.enabled = !0;
    }, this.orientation = (_t$orientation = t.orientation) != null ? _t$orientation : oe.Vertical, this.proportionalLayout = (_t$proportionalLayout = t.proportionalLayout) != null ? _t$proportionalLayout : !0, this.getSashOrthogonalSize = t.getSashOrthogonalSize, i && (this.onDidChange = i), s && (this.onDidDragStart = s), n && (this.onDidDragEnd = n), this.sashContainer = document.createElement("div"), this.sashContainer.classList.add("sash-container", D), e.prepend(this.sashContainer), t.descriptor) {
      this.size = t.descriptor.size;
      for (var [_e6, _i] of t.descriptor.views.entries()) {
        var _t = _i.size,
          _s = _i.container,
          _n = _i.view;
        this.addView(_s, _n, _t, _e6, !0);
      }
      this.contentSize = this.viewItems.reduce((e, t) => e + t.size, 0), this.saveProportions();
    }
  }
  addView(e, t, i, s, n) {
    if (s === void 0) {
      s = this.viewItems.length;
    }
    var r;
    r = "number" == typeof i ? i : "split" === i.type ? this.getViewSize(i.index) / 2 : "invisible" === i.type ? {
      cachedVisibleSize: i.cachedVisibleSize
    } : t.minimumSize;
    var o = this.orientation === oe.Vertical ? new ve(e, t, r) : new pe(e, t, r);
    if (this.viewItems.splice(s, 0, o), this.viewItems.length > 1) {
      var _e7 = this.orientation === oe.Vertical ? new ue(this.sashContainer, {
          getHorizontalSashTop: e => this.getSashPosition(e),
          getHorizontalSashWidth: this.getSashOrthogonalSize
        }, {
          orientation: oe.Horizontal
        }) : new ue(this.sashContainer, {
          getVerticalSashLeft: e => this.getSashPosition(e),
          getVerticalSashHeight: this.getSashOrthogonalSize
        }, {
          orientation: oe.Vertical
        }),
        _t2 = this.orientation === oe.Vertical ? t => ({
          sash: _e7,
          start: t.startY,
          current: t.currentY
        }) : t => ({
          sash: _e7,
          start: t.startX,
          current: t.currentX
        });
      _e7.on("start", e => {
        var _this$onDidDragStart;
        this.emit("sashDragStart"), this.onSashStart(_t2(e));
        var i = this.viewItems.map(e => e.size);
        (_this$onDidDragStart = this.onDidDragStart) == null || _this$onDidDragStart.call(this, i);
      }), _e7.on("change", e => this.onSashChange(_t2(e))), _e7.on("end", () => {
        var _this$onDidDragEnd;
        this.emit("sashDragEnd"), this.onSashEnd(this.sashItems.findIndex(t => t.sash === _e7));
        var t = this.viewItems.map(e => e.size);
        (_this$onDidDragEnd = this.onDidDragEnd) == null || _this$onDidDragEnd.call(this, t);
      }), _e7.on("reset", () => {
        var t = this.sashItems.findIndex(t => t.sash === _e7),
          i = K(t, -1, -1),
          s = K(t + 1, this.viewItems.length),
          n = this.findFirstSnapIndex(i),
          r = this.findFirstSnapIndex(s);
        ("number" != typeof n || this.viewItems[n].visible) && ("number" != typeof r || this.viewItems[r].visible) && this.emit("sashreset", t);
      });
      var _i2 = {
        sash: _e7
      };
      this.sashItems.splice(s - 1, 0, _i2);
    }
    n || this.relayout(), n || "number" == typeof i || "distribute" !== i.type || this.distributeViewSizes();
  }
  removeView(e, t) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    var i = this.viewItems.splice(e, 1)[0].view;
    if (this.viewItems.length >= 1) {
      var _t3 = Math.max(e - 1, 0);
      this.sashItems.splice(_t3, 1)[0].sash.dispose();
    }
    return this.relayout(), t && "distribute" === t.type && this.distributeViewSizes(), i;
  }
  moveView(e, t, i) {
    var s = this.getViewCachedVisibleSize(t),
      n = void 0 === s ? this.getViewSize(t) : ce.Invisible(s),
      r = this.removeView(t);
    this.addView(e, r, n, i);
  }
  getViewCachedVisibleSize(e) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    return this.viewItems[e].cachedVisibleSize;
  }
  layout(e) {
    if (e === void 0) {
      e = this.size;
    }
    var t = Math.max(this.size, this.contentSize);
    if (this.size = e, this.proportions) for (var _t4 = 0; _t4 < this.viewItems.length; _t4++) {
      var _i3 = this.viewItems[_t4];
      _i3.size = x(Math.round(this.proportions[_t4] * e), _i3.minimumSize, _i3.maximumSize);
    } else {
      var _i4 = K(0, this.viewItems.length),
        _s2 = _i4.filter(e => this.viewItems[e].priority === de.Low),
        _n2 = _i4.filter(e => this.viewItems[e].priority === de.High);
      this.resize(this.viewItems.length - 1, e - t, void 0, _s2, _n2);
    }
    this.distributeEmptySpace(), this.layoutViews();
  }
  resizeView(e, t) {
    if (e < 0 || e >= this.viewItems.length) return;
    var i = K(0, this.viewItems.length).filter(t => t !== e),
      s = [...i.filter(e => this.viewItems[e].priority === de.Low), e],
      n = i.filter(e => this.viewItems[e].priority === de.High),
      r = this.viewItems[e];
    t = Math.round(t), t = x(t, r.minimumSize, Math.min(r.maximumSize, this.size)), r.size = t, this.relayout(s, n);
  }
  resizeViews(e) {
    for (var _t5 = 0; _t5 < e.length; _t5++) {
      var _i5 = this.viewItems[_t5];
      var _s3 = e[_t5];
      _s3 = Math.round(_s3), _s3 = x(_s3, _i5.minimumSize, Math.min(_i5.maximumSize, this.size)), _i5.size = _s3;
    }
    this.contentSize = this.viewItems.reduce((e, t) => e + t.size, 0), this.saveProportions(), this.layout(this.size);
  }
  getViewSize(e) {
    return e < 0 || e >= this.viewItems.length ? -1 : this.viewItems[e].size;
  }
  isViewVisible(e) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    return this.viewItems[e].visible;
  }
  setViewVisible(e, t) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    this.viewItems[e].setVisible(t), this.distributeEmptySpace(e), this.layoutViews(), this.saveProportions();
  }
  distributeViewSizes() {
    var e = [];
    var t = 0;
    for (var _i6 of this.viewItems) _i6.maximumSize - _i6.minimumSize > 0 && (e.push(_i6), t += _i6.size);
    var i = Math.floor(t / e.length);
    for (var _t6 of e) _t6.size = x(i, _t6.minimumSize, _t6.maximumSize);
    var s = K(0, this.viewItems.length),
      n = s.filter(e => this.viewItems[e].priority === de.Low),
      r = s.filter(e => this.viewItems[e].priority === de.High);
    this.relayout(n, r);
  }
  dispose() {
    this.sashItems.forEach(e => e.sash.dispose()), this.sashItems = [], this.sashContainer.remove();
  }
  relayout(e, t) {
    var i = this.viewItems.reduce((e, t) => e + t.size, 0);
    this.resize(this.viewItems.length - 1, this.size - i, void 0, e, t), this.distributeEmptySpace(), this.layoutViews(), this.saveProportions();
  }
  onSashStart(_ref2) {
    var {
      sash: e,
      start: t
    } = _ref2;
    var i = this.sashItems.findIndex(t => t.sash === e);
    (e => {
      var t = this.viewItems.map(e => e.size);
      var s,
        n,
        r = Number.NEGATIVE_INFINITY,
        o = Number.POSITIVE_INFINITY;
      var a = K(i, -1, -1),
        h = K(i + 1, this.viewItems.length),
        l = a.reduce((e, i) => e + (this.viewItems[i].minimumSize - t[i]), 0),
        u = a.reduce((e, i) => e + (this.viewItems[i].viewMaximumSize - t[i]), 0),
        c = 0 === h.length ? Number.POSITIVE_INFINITY : h.reduce((e, i) => e + (t[i] - this.viewItems[i].minimumSize), 0),
        m = 0 === h.length ? Number.NEGATIVE_INFINITY : h.reduce((e, i) => e + (t[i] - this.viewItems[i].viewMaximumSize), 0);
      r = Math.max(l, m), o = Math.min(c, u);
      var d = this.findFirstSnapIndex(a),
        f = this.findFirstSnapIndex(h);
      if ("number" == typeof d) {
        var _e8 = this.viewItems[d],
          _t7 = Math.floor(_e8.viewMinimumSize / 2);
        s = {
          index: d,
          limitDelta: _e8.visible ? r - _t7 : r + _t7,
          size: _e8.size
        };
      }
      if ("number" == typeof f) {
        var _e9 = this.viewItems[f],
          _t8 = Math.floor(_e9.viewMinimumSize / 2);
        n = {
          index: f,
          limitDelta: _e9.visible ? o + _t8 : o - _t8,
          size: _e9.size
        };
      }
      this.sashDragState = {
        start: e,
        current: e,
        index: i,
        sizes: t,
        minDelta: r,
        maxDelta: o,
        snapBefore: s,
        snapAfter: n
      };
    })(t);
  }
  onSashChange(_ref3) {
    var {
      current: e
    } = _ref3;
    var {
      index: t,
      start: i,
      sizes: s,
      minDelta: n,
      maxDelta: r,
      snapBefore: o,
      snapAfter: a
    } = this.sashDragState;
    this.sashDragState.current = e;
    var h = e - i;
    this.resize(t, h, s, void 0, void 0, n, r, o, a), this.distributeEmptySpace(), this.layoutViews();
  }
  getSashPosition(e) {
    var t = 0;
    for (var _i7 = 0; _i7 < this.sashItems.length; _i7++) if (t += this.viewItems[_i7].size, this.sashItems[_i7].sash === e) return t;
    return 0;
  }
  resize(e, t, i, s, n, r, o, a, h) {
    if (i === void 0) {
      i = this.viewItems.map(e => e.size);
    }
    if (r === void 0) {
      r = Number.NEGATIVE_INFINITY;
    }
    if (o === void 0) {
      o = Number.POSITIVE_INFINITY;
    }
    if (e < 0 || e >= this.viewItems.length) return 0;
    var l = K(e, -1, -1),
      u = K(e + 1, this.viewItems.length);
    if (n) for (var _e0 of n) J(l, _e0), J(u, _e0);
    if (s) for (var _e1 of s) U(l, _e1), U(u, _e1);
    var c = l.map(e => this.viewItems[e]),
      m = l.map(e => i[e]),
      d = u.map(e => this.viewItems[e]),
      f = u.map(e => i[e]),
      p = l.reduce((e, t) => e + (this.viewItems[t].minimumSize - i[t]), 0),
      v = l.reduce((e, t) => e + (this.viewItems[t].maximumSize - i[t]), 0),
      S = 0 === u.length ? Number.POSITIVE_INFINITY : u.reduce((e, t) => e + (i[t] - this.viewItems[t].minimumSize), 0),
      z = 0 === u.length ? Number.NEGATIVE_INFINITY : u.reduce((e, t) => e + (i[t] - this.viewItems[t].maximumSize), 0),
      w = Math.max(p, z, r),
      g = Math.min(S, v, o);
    var y = !1;
    if (a) {
      var _e10 = this.viewItems[a.index],
        _i8 = t >= a.limitDelta;
      y = _i8 !== _e10.visible, _e10.setVisible(_i8, a.size);
    }
    if (!y && h) {
      var _e11 = this.viewItems[h.index],
        _i9 = t < h.limitDelta;
      y = _i9 !== _e11.visible, _e11.setVisible(_i9, h.size);
    }
    if (y) return this.resize(e, t, i, s, n, r, o);
    for (var _e12 = 0, _i0 = t = x(t, w, g); _e12 < c.length; _e12++) {
      var _t9 = c[_e12],
        _s4 = x(m[_e12] + _i0, _t9.minimumSize, _t9.maximumSize);
      _i0 -= _s4 - m[_e12], _t9.size = _s4;
    }
    for (var _e13 = 0, _i1 = t; _e13 < d.length; _e13++) {
      var _t0 = d[_e13],
        _s5 = x(f[_e13] - _i1, _t0.minimumSize, _t0.maximumSize);
      _i1 += _s5 - f[_e13], _t0.size = _s5;
    }
    return t;
  }
  distributeEmptySpace(e) {
    var t = this.viewItems.reduce((e, t) => e + t.size, 0);
    var i = this.size - t;
    var s = K(0, this.viewItems.length),
      n = [],
      r = s.filter(e => this.viewItems[e].priority === de.Low),
      o = s.filter(e => this.viewItems[e].priority === de.Normal),
      a = s.filter(e => this.viewItems[e].priority === de.High);
    n.push(...a, ...o, ...r), "number" == typeof e && U(n, e);
    for (var _e14 = 0; 0 !== i && _e14 < n.length; _e14++) {
      var _t1 = this.viewItems[n[_e14]],
        _s6 = x(_t1.size + i, _t1.minimumSize, _t1.maximumSize);
      i -= _s6 - _t1.size, _t1.size = _s6;
    }
  }
  layoutViews() {
    var _this$onDidChange;
    this.contentSize = this.viewItems.reduce((e, t) => e + t.size, 0);
    var e = 0;
    for (var _t10 of this.viewItems) _t10.layout(e), e += _t10.size;
    (_this$onDidChange = this.onDidChange) != null && _this$onDidChange.call(this, this.viewItems.map(e => e.size)), this.sashItems.forEach(e => e.sash.layout()), this.updateSashEnablement();
  }
  saveProportions() {
    this.proportionalLayout && this.contentSize > 0 && (this.proportions = this.viewItems.map(e => e.size / this.contentSize));
  }
  updateSashEnablement() {
    var e = !1;
    var t = this.viewItems.map(t => e = t.size - t.minimumSize > 0 || e);
    e = !1;
    var i = this.viewItems.map(t => e = t.maximumSize - t.size > 0 || e),
      s = [...this.viewItems].reverse();
    e = !1;
    var n = s.map(t => e = t.size - t.minimumSize > 0 || e).reverse();
    e = !1;
    var r = s.map(t => e = t.maximumSize - t.size > 0 || e).reverse();
    var o = 0;
    for (var _e15 = 0; _e15 < this.sashItems.length; _e15++) {
      var {
        sash: _s7
      } = this.sashItems[_e15];
      o += this.viewItems[_e15].size;
      var _a = !(t[_e15] && r[_e15 + 1]),
        _h = !(i[_e15] && n[_e15 + 1]);
      if (_a && _h) {
        var _i10 = K(_e15, -1, -1),
          _r = K(_e15 + 1, this.viewItems.length),
          _a2 = this.findFirstSnapIndex(_i10),
          _h2 = this.findFirstSnapIndex(_r),
          _l = "number" == typeof _a2 && !this.viewItems[_a2].visible,
          _u = "number" == typeof _h2 && !this.viewItems[_h2].visible;
        _l && n[_e15] && (o > 0 || this.startSnappingEnabled) ? _s7.state = ae.Minimum : _u && t[_e15] && (o < this.contentSize || this.endSnappingEnabled) ? _s7.state = ae.Maximum : _s7.state = ae.Disabled;
      } else _s7.state = _a && !_h ? ae.Minimum : !_a && _h ? ae.Maximum : ae.Enabled;
    }
  }
  findFirstSnapIndex(e) {
    for (var _t11 of e) {
      var _e16 = this.viewItems[_t11];
      if (_e16.visible && _e16.snap) return _t11;
    }
    for (var _t12 of e) {
      var _e17 = this.viewItems[_t12];
      if (_e17.visible && _e17.maximumSize - _e17.minimumSize > 0) return;
      if (!_e17.visible && _e17.snap) return _t12;
    }
  }
}
class ze {
  constructor(e) {
    this.size = void 0, this.size = e;
  }
  getPreferredSize() {
    return this.size;
  }
}
class we {
  constructor(e, t) {
    this.proportion = void 0, this.layoutService = void 0, this.proportion = e, this.layoutService = t;
  }
  getPreferredSize() {
    return this.proportion * this.layoutService.getSize();
  }
}
class ge {
  getPreferredSize() {}
}
class ye {
  get preferredSize() {
    return this.layoutStrategy.getPreferredSize();
  }
  set preferredSize(e) {
    if ("number" == typeof e) this.layoutStrategy = new ze(e);else if ("string" == typeof e) {
      var _t13 = e.trim();
      if (B(_t13, "%")) {
        var _e18 = Number(_t13.slice(0, -1)) / 100;
        this.layoutStrategy = new we(_e18, this.layoutService);
      } else if (B(_t13, "px")) {
        var _e19 = Number(_t13.slice(0, -2)) / 100;
        this.layoutStrategy = new ze(_e19);
      } else if ("number" == typeof Number.parseFloat(_t13)) {
        var _e20 = Number.parseFloat(_t13);
        this.layoutStrategy = new ze(_e20);
      } else this.layoutStrategy = new ge();
    } else this.layoutStrategy = new ge();
  }
  constructor(e, t) {
    var _t$priority;
    if (this.minimumSize = 0, this.maximumSize = Number.POSITIVE_INFINITY, this.element = void 0, this.priority = void 0, this.snap = void 0, this.layoutService = void 0, this.layoutStrategy = void 0, this.layoutService = e, this.element = t.element, this.minimumSize = "number" == typeof t.minimumSize ? t.minimumSize : 30, this.maximumSize = "number" == typeof t.maximumSize ? t.maximumSize : Number.POSITIVE_INFINITY, "number" == typeof t.preferredSize) this.layoutStrategy = new ze(t.preferredSize);else if ("string" == typeof t.preferredSize) {
      var _e21 = t.preferredSize.trim();
      if (B(_e21, "%")) {
        var _t14 = Number(_e21.slice(0, -1)) / 100;
        this.layoutStrategy = new we(_t14, this.layoutService);
      } else if (B(_e21, "px")) {
        var _t15 = Number(_e21.slice(0, -2));
        this.layoutStrategy = new ze(_t15);
      } else if ("number" == typeof Number.parseFloat(_e21)) {
        var _t16 = Number.parseFloat(_e21);
        this.layoutStrategy = new ze(_t16);
      } else this.layoutStrategy = new ge();
    } else this.layoutStrategy = new ge();
    this.priority = (_t$priority = t.priority) != null ? _t$priority : de.Normal, this.snap = "boolean" == typeof t.snap && t.snap;
  }
  layout(e) {}
}
function be(e) {
  return void 0 !== e.minSize || void 0 !== e.maxSize || void 0 !== e.preferredSize || void 0 !== e.priority || void 0 !== e.visible;
}
var Ie = o((_ref4, s) => {
  var {
    className: t,
    children: i
  } = _ref4;
  return e.createElement("div", {
    ref: s,
    className: S("split-view-view", M, t)
  }, i);
});
Ie.displayName = "Allotment.Pane";
var xe = o((_ref5, _) => {
  var {
    children: r,
    className: o,
    id: l,
    maxSize: u = 1 / 0,
    minSize: c = 30,
    proportionalLayout: m = !0,
    separator: d = !0,
    sizes: f,
    defaultSizes: p = f,
    snap: v = !1,
    vertical: z = !1,
    onChange: w,
    onReset: g,
    onVisibleChange: b,
    onDragStart: I,
    onDragEnd: x
  } = _ref5;
  var V = i(null),
    N = i([]),
    D = i(new Map()),
    M = i(null),
    A = i(new Map()),
    j = i(new $()),
    F = i([]),
    [Y, B] = t(!1);
  "production" !== process.env.NODE_ENV && f && console.warn("Prop sizes is deprecated. Please use defaultSizes instead.");
  var R = a(() => e.Children.toArray(r).filter(e.isValidElement), [r]),
    W = n(e => {
      var _F$current, _M$current;
      var t = (_F$current = F.current) == null ? void 0 : _F$current[e];
      return "number" == typeof (t == null ? void 0 : t.preferredSize) && ((_M$current = M.current) != null && _M$current.resizeView(e, Math.round(t.preferredSize)), !0);
    }, []);
  return h(_, () => ({
    reset: () => {
      if (g) g();else {
        var _M$current2;
        (_M$current2 = M.current) == null || _M$current2.distributeViewSizes();
        for (var _e22 = 0; _e22 < F.current.length; _e22++) W(_e22);
      }
    },
    resize: e => {
      var _M$current3;
      (_M$current3 = M.current) == null || _M$current3.resizeViews(e);
    }
  })), k(() => {
    var e = !0;
    p && A.current.size !== p.length && (e = !1, console.warn("Expected " + p.length + " children based on defaultSizes but found " + A.current.size)), e && p && (N.current = R.map(e => e.key));
    var t = _extends({
      orientation: z ? oe.Vertical : oe.Horizontal,
      proportionalLayout: m
    }, e && p && {
      descriptor: {
        size: p.reduce((e, t) => e + t, 0),
        views: p.map((e, t) => {
          var _i$minSize, _i$maxSize, _i$priority, _i$snap;
          var i = D.current.get(N.current[t]),
            s = new ye(j.current, _extends({
              element: document.createElement("div"),
              minimumSize: (_i$minSize = i == null ? void 0 : i.minSize) != null ? _i$minSize : c,
              maximumSize: (_i$maxSize = i == null ? void 0 : i.maxSize) != null ? _i$maxSize : u,
              priority: (_i$priority = i == null ? void 0 : i.priority) != null ? _i$priority : de.Normal
            }, (i == null ? void 0 : i.preferredSize) && {
              preferredSize: i == null ? void 0 : i.preferredSize
            }, {
              snap: (_i$snap = i == null ? void 0 : i.snap) != null ? _i$snap : v
            }));
          return F.current.push(s), {
            container: [...A.current.values()][t],
            size: e,
            view: s
          };
        })
      }
    });
    M.current = new Se(V.current, t, w, I, x), M.current.on("sashDragStart", () => {
      var _V$current;
      (_V$current = V.current) == null || _V$current.classList.add("split-view-sash-dragging");
    }), M.current.on("sashDragEnd", () => {
      var _V$current2;
      (_V$current2 = V.current) == null || _V$current2.classList.remove("split-view-sash-dragging");
    }), M.current.on("sashchange", e => {
      if (b && M.current) {
        var _e23 = R.map(e => e.key);
        for (var _t17 = 0; _t17 < _e23.length; _t17++) {
          var _i11 = D.current.get(_e23[_t17]);
          void 0 !== (_i11 == null ? void 0 : _i11.visible) && _i11.visible !== M.current.isViewVisible(_t17) && b(_t17, M.current.isViewVisible(_t17));
        }
      }
    }), M.current.on("sashreset", e => {
      if (g) g();else {
        var _M$current4;
        if (W(e)) return;
        if (W(e + 1)) return;
        (_M$current4 = M.current) == null || _M$current4.distributeViewSizes();
      }
    });
    var i = M.current;
    return () => {
      i.dispose();
    };
  }, []), k(() => {
    if (Y) {
      var _e24 = R.map(e => e.key),
        _t18 = [...N.current],
        _i12 = _e24.filter(e => !N.current.includes(e)),
        _s8 = _e24.filter(e => N.current.includes(e)),
        _n3 = N.current.map(t => !_e24.includes(t));
      for (var _e25 = _n3.length - 1; _e25 >= 0; _e25--) {
        var _M$current5;
        _n3[_e25] && ((_M$current5 = M.current) != null && _M$current5.removeView(_e25), _t18.splice(_e25, 1), F.current.splice(_e25, 1));
      }
      var _loop = function _loop(_s9) {
        var _i$minSize2, _i$maxSize2, _i$priority2, _i$snap2, _M$current7;
        var i = D.current.get(_s9),
          n = new ye(j.current, _extends({
            element: document.createElement("div"),
            minimumSize: (_i$minSize2 = i == null ? void 0 : i.minSize) != null ? _i$minSize2 : c,
            maximumSize: (_i$maxSize2 = i == null ? void 0 : i.maxSize) != null ? _i$maxSize2 : u,
            priority: (_i$priority2 = i == null ? void 0 : i.priority) != null ? _i$priority2 : de.Normal
          }, (i == null ? void 0 : i.preferredSize) && {
            preferredSize: i == null ? void 0 : i.preferredSize
          }, {
            snap: (_i$snap2 = i == null ? void 0 : i.snap) != null ? _i$snap2 : v
          }));
        (_M$current7 = M.current) != null && _M$current7.addView(A.current.get(_s9), n, ce.Distribute, _e24.findIndex(e => e === _s9)), _t18.splice(_e24.findIndex(e => e === _s9), 0, _s9), F.current.splice(_e24.findIndex(e => e === _s9), 0, n);
      };
      for (var _s9 of _i12) {
        _loop(_s9);
      }
      for (; !y(_e24, _t18);) {
        var _loop2 = function _loop2(_s0) {
          var e = _t18.findIndex(e => e === _s0);
          if (e !== _i13) {
            var _M$current6;
            (_M$current6 = M.current) == null || _M$current6.moveView(A.current.get(_s0), e, _i13);
            var _n4 = _t18[e];
            _t18.splice(e, 1), _t18.splice(_i13, 0, _n4);
            return 1; // break
          }
        };
        for (var [_i13, _s0] of _e24.entries()) {
          if (_loop2(_s0)) break;
        }
      }
      var _loop3 = function _loop3(_t19) {
        var _M$current8;
        var i = _e24.findIndex(e => e === _t19),
          s = F.current[i].preferredSize;
        void 0 !== s && ((_M$current8 = M.current) == null ? void 0 : _M$current8.resizeView(i, s));
      };
      for (var _t19 of _i12) {
        _loop3(_t19);
      }
      var _loop4 = function _loop4(_t20) {
        var _M$current9, _M$current0;
        var i = D.current.get(_t20),
          s = _e24.findIndex(e => e === _t20);
        i && be(i) && void 0 !== i.visible && ((_M$current9 = M.current) == null ? void 0 : _M$current9.isViewVisible(s)) !== i.visible && ((_M$current0 = M.current) == null ? void 0 : _M$current0.setViewVisible(s, i.visible));
      };
      for (var _t20 of [..._i12, ..._s8]) {
        _loop4(_t20);
      }
      var _loop5 = function _loop5(_t21) {
        var i = D.current.get(_t21),
          s = _e24.findIndex(e => e === _t21);
        if (i && be(i)) {
          var _M$current1;
          void 0 !== i.preferredSize && F.current[s].preferredSize !== i.preferredSize && (F.current[s].preferredSize = i.preferredSize);
          var _e26 = !1;
          void 0 !== i.minSize && F.current[s].minimumSize !== i.minSize && (F.current[s].minimumSize = i.minSize, _e26 = !0), void 0 !== i.maxSize && F.current[s].maximumSize !== i.maxSize && (F.current[s].maximumSize = i.maxSize, _e26 = !0), _e26 && ((_M$current1 = M.current) == null ? void 0 : _M$current1.layout());
        }
      };
      for (var _t21 of _s8) {
        _loop5(_t21);
      }
      (_i12.length > 0 || _n3.length > 0) && (N.current = _e24);
    }
  }, [R, Y, u, c, v]), s(() => {
    M.current && (M.current.onDidChange = w);
  }, [w]), s(() => {
    M.current && (M.current.onDidDragStart = I);
  }, [I]), s(() => {
    M.current && (M.current.onDidDragEnd = x);
  }, [x]), E({
    ref: V,
    onResize: _ref6 => {
      var _M$current10;
      var {
        width: e,
        height: t
      } = _ref6;
      e && t && ((_M$current10 = M.current) != null && _M$current10.layout(z ? t : e), j.current.setSize(z ? t : e), B(!0));
    }
  }), s(() => {
    H && _e(20);
  }, []), e.createElement("div", {
    ref: V,
    className: S("split-view", z ? "split-view-vertical" : "split-view-horizontal", {
      "split-view-separator-border": d
    }, L, z ? P : T, {
      [C]: d
    }, o),
    id: l
  }, e.createElement("div", {
    className: S("split-view-container", O)
  }, e.Children.toArray(r).map(t => {
    if (!e.isValidElement(t)) return null;
    var i = t.key;
    return "Allotment.Pane" === t.type.displayName ? (D.current.set(i, t.props), e.cloneElement(t, {
      key: i,
      ref: e => {
        var s = t.ref;
        s && (s.current = e), e ? A.current.set(i, e) : A.current.delete(i);
      }
    })) : e.createElement(Ie, {
      key: i,
      ref: e => {
        e ? A.current.set(i, e) : A.current.delete(i);
      }
    }, t);
  })));
});
function _e(e) {
  var t = x(e, 4, 20),
    i = x(e, 1, 8);
  document.documentElement.style.setProperty("--sash-size", t + "px"), document.documentElement.style.setProperty("--sash-hover-size", i + "px"), function (e) {
    he = e, le.emit("onDidChangeGlobalSize", e);
  }(t);
}
xe.displayName = "Allotment";
var Ve = Object.assign(xe, {
  Pane: Ie
});
export { Ve as Allotment, de as LayoutPriority, _e as setSashSize };
